---
title: "16S Data Analysis"
author: 
  - name: Lan Huong Nguyen
    affiliation: Institute for Computational and Mathematical Engineering, Stanford University, CA 94305
keywords: microbiome, metatranscriptomics.
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{Gene expression time course data analysis}
  %\VignetteEngine{knitr::rmarkdown}
---



System Setup {#setup .unnumbered}
========================================

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("rmarkdown")
options(width = 200, stringsAsFactors = FALSE) 
knitr::opts_chunk$set(
  message = FALSE, error = FALSE, warning = FALSE, 
  fig.width = 8, fig.height = 6,
  fig.path = "../figs/holmes_group_presentation/", 
  dev='png') 
```

```{r setup}
rm(list = ls())
library("DESeq2")
library("viridis")
library("RColorBrewer")
library("tidyverse")
library("readxl")
library("phyloseq")


theme_set(theme_bw())
theme_update(
  text = element_text(15),
  panel.border = element_rect(size = 0.5),
  panel.grid = element_blank(),
  axis.ticks = element_blank(),
  legend.title = element_text(size = 12),
  legend.text = element_text(size = 12),
  legend.position = "bottom",
  strip.background = element_blank()
)
datadir <- "/scratch/PI/sph/resilience/"
curdir <- getwd()
```


```{r}
plot_projection <- function(data, xname, yname, labname = "Subject", size = 3,
                            color = NULL, eigs = NULL, ...){
  if(all(!is.null(color), color %in% colnames(data))){
    plt <- ggplot(data, aes_string(x = xname, y = yname, color = color)) 
  } else {
    plt <- ggplot(data, aes_string(x = xname, y = yname)) 
  }
  if(all(!is.null(labname), labname %in% colnames(data))) {
    if(size %in% colnames(data)) {
      plt <- plt + geom_text(aes_string(label = labname, size = size), ...)
    } else {
      plt <- plt + geom_text(aes_string(label = labname), size = size, ...)
    }
  } else {
    if(size %in% colnames(data)) {
      plt <- plt + geom_point(aes_string(size = size), ...)
    } else {
      plt <- plt + geom_point(size = size, ...)
    }
  }
  if(!is.null(eigs)){
    var.explained <- round(100 * eigs/sum(eigs), 2)
    axis1 <- as.numeric(gsub("\\D", "", xname))
    axis2 <- as.numeric(gsub("\\D", "", yname))
    plt <- plt + 
      labs(
        x = sprintf("Axis%d [%s%% variance]", axis1, var.explained[axis1]),
        y = sprintf("Axis%d [%s%% variance]", axis2, var.explained[axis2])
      ) +
      coord_fixed(sqrt( eigs[axis2]/eigs[axis1])) 
  }
  return(plt + theme(text = element_text(size = 20))) 
}
```


Sample Information {#sample-info .unnumbered}
========================================


```{r, eval = FALSE}
# loading sample info
smpinfo_file <- file.path(datadir, "sample_info/Mapping_File_10June2018.xlsx") 

# Measurement Information
meas <- read_xlsx(smpinfo_file, sheet = "Meas")
colnames(meas) <- meas[1, ]
meas <- meas[-1, ] %>%
  rename(Samp_ID = SampID) %>%
  select(Meas_ID:Meas_Type)
```

```{r, eval = FALSE}
# Sample Information
interv_levs <- c("NoInterv", 
                 "PreDiet", "MidDiet", "PostDiet", 
                 "PreCC", "MidCC", "PostCC", 
                 "PreAbx", "MidAbx", "PostAbx", "UnpAbx")

samp <- read_xlsx(smpinfo_file, "Samp", skip = 1)  %>%
  filter(Samp_Type != "ExtrCont") %>%
  mutate_at(
    .vars = c("Diet_Interval", "CC_Interval", "Abx_Interval"),
    .funs = function(x) {
      x <- ifelse(x == "Midabx","MidAbx", x)
      x <- ifelse(x == "NA", "NoInterv", x)
      x <- factor(x, interv_levs)
      return(x)
    }
  ) %>%
  mutate_at(
    .vars = vars(contains("Date")),
    .funs = function(x) {
      as.Date(as.numeric(x), origin="1899-12-30")
    }
  ) %>%
  mutate(
    Interval = paste0(Diet_Interval, "_", CC_Interval, "_", Abx_Interval),
    Interval = ifelse(grepl("MidAbx", Interval), "MidAbx", Interval),
    Interval = ifelse(grepl("MidDiet", Interval), "MidDiet", Interval),
    Interval = ifelse(grepl("MidCC", Interval), "MidCC", Interval),
    Interval = ifelse(!grepl("Post", Interval) & !grepl("Mid", Interval) &
                  !grepl("UnpAbx", Interval), "NoInterv", Interval),
    Interval = gsub("NoInterv_", "", Interval),
    Interval = gsub("_NoInterv", "", Interval),
    Interval = factor(Interval, unique(Interval))
  ) %>%
  select(Samp_ID:Abx_RelDay, Interval) 
head(samp)
```

```{r, eval = FALSE}
# Subject Data
subj_data <- read_excel(
  smpinfo_file, sheet = "SubjNoRagged", skip = 1, 
  col_types = c("text", rep("date", 4), rep("guess", 864))
  )  %>%
  mutate(
    Group = sapply(1:nrow(.), function(i) {
      group <- c()
      if(!is.na(Diet_StartDate[i])) group <- c(group, "Diet")
      if(!is.na(CC_Date[i])) group <- c(group, "CC")
      if(!is.na(Abx_StartDate[i])) group <- c(group, "Abx")
      if(length(group) == 0) group <- "NoIntv"
      return(paste(group, collapse = "_"))
    })
    ) %>%
  select(Subject, Group, First_Sample_Date, CC_Date, Diet_StartDate, 
         Abx_StartDate, Age:BMI)
head(subj_data)
```

```{r eval = FALSE}
meas_info <- meas %>%
  left_join(samp) %>%
  left_join(subj_data) 
write.csv(meas_info, "./meas_info_10June2018.csv")
```

```{r}
meas_info <- read.csv("./meas_info_10June2018.csv", row.names = 1) %>%
  mutate_at(.vars = vars(contains("Date")), 
            .funs = function(x) {as.Date(as.character(x), "%Y-%m-%d")}) 
head(meas_info)
```

```{r}
start_dates <- meas_info %>%
   group_by(Subject, First_Sample_Date) %>%
    summarise(Real_First_Sample_Date = min(Samp_Date, na.rm = TRUE))
start_dates %>% filter(First_Sample_Date != Real_First_Sample_Date)
```

Fix one incorrect first sample date:

```{r eval = FALSE}
meas_info <- meas_info %>%
  left_join(start_dates) %>%
  mutate(DaysFromStart = as.numeric(Samp_Date - Real_First_Sample_Date ))
write.csv(meas_info, "./meas_info_10June2018.csv")
```


```{r}
smp_modality_count <- meas_info %>%
  select(Meas_ID, Samp_ID, Meas_Type) %>%
  group_by(Samp_ID) %>%
  summarise(
    modalities = paste0(unique(Meas_Type), collapse = "_"),
    modality_count = n()
  ) %>%
  mutate(
    modalities = ifelse(modalities == "MetaT_MetaG_16S", "16S_MetaG_MetaT", modalities)
  )
  
df2plot <- meas_info %>%
  left_join(smp_modality_count) %>% 
  select(Meas_ID, Samp_ID, Subject, Samp_Date, Meas_Type, 
         Group, Interval, modalities, modality_count) %>%
  mutate(Samp_Date = as.Date(Samp_Date)) %>%
  filter(!grepl("NA", Subject))
```

```{r exp-design, fig.width=10, fig.height=8}
new_intv_levs <- unique(meas_info$Interval)
colors <- colorRampPalette(brewer.pal(8, "Set3"))(length(new_intv_levs)-2)
colors <- c("grey57", "red", colors)
names(colors) <- c("NoInterv", "UnpAbx", 
                   setdiff(new_intv_levs, c("NoInterv", "UnpAbx")))
ggplot(df2plot %>% filter(Group != "Abx")) +
  geom_point(
      pch = 124,
      aes(y = Subject, x = Samp_Date, color = Interval,
          fill = Interval, size = modality_count)
    ) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    scale_size_continuous(range = c(5, 10)) +
    theme(legend.direction = "horizontal", legend.box = "vertical",
          strip.text.y = element_text(angle=0)) +
    guides(color = guide_legend(override.aes = list(size=5)),
           size = guide_legend(override.aes = list(color="black")))
```


```{r}
rm("colors", "df2plot", "smp_modality_count", "new_intv_levs")
```


Generate Phyloseq Object {#generate-phyloseq .unnumbered}
========================================

First we need to generate a new phyloseq object with samples
from two new batches.

We read the data stored as a `phyloseq` object from batches 1-4, 
then 5-6.


```{r get-physeq-batches-1-4, eval = FALSE}
# Bacthes 1-4
first_batches_dir <- file.path(datadir, "16S/16S_dada2_results/")

psRDP <- readRDS(file.path(first_batches_dir, "phyloseq_RDP.rds"))
psRDP

psRDPMult <- readRDS(file.path(first_batches_dir, "phyloseq_RDP_multispecies.rds"))
psRDPMult

psSilva <- readRDS(file.path(first_batches_dir, "phyloseq_Silva.rds"))
psSilva

psSilvaMult <- readRDS(file.path(first_batches_dir, "phyloseq_Silva_multispecies.rds"))
psSilvaMult
```

```{r get-physeq-batches-5-6, eval = FALSE}
# Bacthes 5-6
second_batches_dir <- file.path(datadir, "16S/Les_Results/16S_5-6_DADA2_output")

psRDP2 <- readRDS(file.path(second_batches_dir, "phyloseq_RDP.rds"))
psRDP2

psRDPMult2 <- readRDS(file.path(second_batches_dir, "phyloseq_RDP_multispecies.rds"))
psRDPMult2

psSilva2 <- readRDS(file.path(second_batches_dir, "phyloseq_Silva.rds"))
psSilva2

psSilvaMult2 <- readRDS(file.path(second_batches_dir, "phyloseq_Silva_multispecies.rds"))
psSilvaMult2
```


```{r, eval = FALSE}
psRDP_merge <- merge_phyloseq(psRDP, psRDP2)
psRDPMult_merge <- merge_phyloseq(psRDPMult, psRDPMult2)
psSilva_merge <- merge_phyloseq(psSilva, psSilva2)
psSilvaMult_merge <- merge_phyloseq(psSilvaMult, psSilvaMult2)
rm(psRDP, psRDP2, psRDPMult, psRDPMult2, psSilva, psSilva2, psSilvaMult, psSilvaMult2)
```


```{r, eval = FALSE}
df <- bind_rows(
  data.frame(DB = "RDP", tax_table(psRDP_merge)@.Data),
  data.frame(DB = "Silva", tax_table(psSilva_merge)@.Data),
  data.frame(DB = "RDP_Mult", tax_table(psRDPMult_merge)@.Data),
  data.frame(DB = "Silva_Mult", tax_table(psSilvaMult_merge)@.Data)
)

# df %>%
#   group_by(DB) %>%
#   summarise_all(.funs = list(
#     function(x) sum(!is.na(x))
#   )
# )

# df %>%
#   group_by(DB) %>%
#   summarise_all(.funs = list(
#     function(x) round(100*sum(!is.na(x))/ntaxa(psRDP_merge), 1)
#   )
# )
```



```{r, eval = FALSE}
newTaxtab <- tax_table(psRDP_merge)
newTaxtab <- cbind(newTaxtab, "SpeciesMult" = tax_table(psRDPMult_merge)[, "Species"])
newTaxtab <- cbind(newTaxtab, "GenusSilva" = tax_table(psSilva_merge)[, "Genus"])
newTaxtab <- cbind(newTaxtab, "SpeciesSilva" = tax_table(psSilva_merge)[, "Species"])
newTaxtab <- cbind(
  newTaxtab, "SpeciesSilvaMult" = tax_table(psSilvaMult_merge)[, "Species"])
newTaxtab <- cbind(newTaxtab, Seq = rownames(newTaxtab))
rownames(newTaxtab) <- paste0("Seq", 1:nrow(newTaxtab))

seqtab <- as(otu_table(psRDP_merge), "matrix")
seqtab <- seqtab[newTaxtab[, "Seq"], ]
rownames(seqtab) <- rownames(newTaxtab) 
```


```{r, eval=FALSE}
ps_smp_names <- colnames(seqtab) 
names(ps_smp_names) <- colnames(seqtab) 

batch_info <- gsub("HMD_", "", ps_smp_names)
batch_info <- gsub("\\_(.*)", "", batch_info)

ps_smp_names <- gsub("\\-(.*)", "", ps_smp_names)
ps_smp_names <- gsub("(.*)\\_M", "M", ps_smp_names)
head(ps_smp_names)
tail(ps_smp_names)
length(ps_smp_names) == length(unique(ps_smp_names))
```

As we see above, the id's are unique in our sample. Now, we check that
all the identifiers are included in the mapping file:

```{r, eval=FALSE}
all(ps_smp_names %in% meas_info$Meas_ID)
```

```{r, eval=FALSE}
ps_smp_names[which(!ps_smp_names %in% meas_info$Meas_ID)]
# "HMD_16S5_SplitLib_Unassigned" 
```


```{r, eval=FALSE}
keep_smp_names <- ps_smp_names[which(ps_smp_names %in% meas_info$Meas_ID)]
batch_info <- batch_info[which(ps_smp_names %in% meas_info$Meas_ID)]
names(batch_info) <- keep_smp_names
seqtab <- seqtab[, names(keep_smp_names)]
colnames(seqtab) <- keep_smp_names
```


We extract the information for the obtained samples:

```{r, eval=FALSE}
samp_16S <- meas_info %>%
  filter(Meas_ID %in% keep_smp_names) %>%
  left_join(data.frame(Meas_ID = names(batch_info), 
                       Sequencing_Run = batch_info))
rownames(samp_16S) <- samp_16S$Meas_ID
head(samp_16S)
```

```{r, eval=FALSE}
ps_combined <- phyloseq(
  otu_table(seqtab, taxa_are_rows = TRUE), 
  sample_data(samp_16S),
  tax_table(newTaxtab))
ps_combined
```

```{r, eval=FALSE}
physeqfile <- file.path(datadir, "16S/phyloseq/perturb_physeq_10Jul18.rds")
saveRDS(ps_combined, file = physeqfile)
```





Preprocess Data {#preprocess .unnumbered}
=========================================



```{r, eval = FALSE}
physeqfile <- file.path(datadir, "16S/phyloseq/perturb_physeq_10Jul18.rds")
ps0 <- readRDS(file = physeqfile)
SMP <- data.frame(ps0@sam_data, stringsAsFactors = FALSE) %>%
  mutate(SampleDepth = sample_sums(ps0))
ps0
```


We will remove samples from subject who did not have sufficient number of
samples sequenced, either because they came from a different experiment or
because the subjects did not complete the study.

```{r, eval = FALSE}
freq <- table(ps0@sam_data$Subject)
remove_subjects <- names(freq)[freq < 10] # "AAA" "AAB" "AAN" "DAC" "DAG"
ps <- subset_samples(ps0, !Subject %in% remove_subjects)
ps
```


Filter taxa {#filter .unnumbered} 
---------------------------------

Next, we will filter out taxa which are not present in samples of at least 
two different subjects.

```{r, eval = FALSE}
minTaxaSubjPrev <- 2

# Compute prevalence across subjects
num_subj_present <- function(physeq, subject_indicator = "Subject") {
  smp <- data.frame(physeq@sam_data, stringsAsFactors = FALSE)
  seqtab <- as(otu_table(physeq), "matrix")
  if(!taxa_are_rows(physeq)) seqtab <- t(seqtab)
  sapply(1:ntaxa(physeq), function(i) {
    subj.present <- smp[[subject_indicator]][seqtab[i, ] > 0]
    length(unique(subj.present))})
}
subjects_prevalence <- num_subj_present(ps)

# Filter taxa present in more than 1 subject
ps <- subset_taxa(ps, subjects_prevalence >= minTaxaSubjPrev)
ps
```

Filter samples {#filter .unnumbered} 
---------------------------------

We then separate the subject data and the negative controls:

```{r, eval = FALSE}
pscontrols <- subset_samples(ps, Subject == "NA_QC" | is.na(Subject))
pscontrols
```

```{r, eval = FALSE}
ps <- subset_samples(ps, Subject != "NA_QC" & !is.na(Subject))
ps
```


We filter the subject samples which have total number of counts less than 10k reads
(over selected ASVs (sequences)).

```{r, eval = FALSE}
minSampleSum <- 1e4
ps <- subset_samples(ps, sample_sums(ps) > minSampleSum)
ps
```

Remove contamination:

```{r}
# TODO
```


```{r, eval = FALSE}
SMP <- data.frame(ps@sam_data, stringsAsFactors = FALSE) %>%
  mutate(SampleDepth = sample_sums(ps))
  
SUBJ <-  SMP %>%
  select(Subject, Group, Age:BMI) %>%
  distinct()
```


```{r, eval = FALSE}
# Transform counts to not "over-weight" highly abundant taxa
ps_asinh <- transform_sample_counts(ps, function(x) {asinh(x)})
```


```{r, eval = FALSE}
# Takes too long
dds <- phyloseq_to_deseq2(ps, design = ~ 1)
dds <- estimateSizeFactors(dds, type = "poscounts")
norm_counts <- counts(dds, normalized = TRUE)

## The following takes too long; and vst() does not work with too few genes with rowmeans > 5
vsd <- varianceStabilizingTransformation(dds)
ps_vst <- ps
otu_table(ps_vst) <- otu_table(assay(vsd), taxa_are_rows = TRUE)
save(list = c("ps", "pscontrols", "ps_asinh", "SMP", "SUBJ", "dds",
              "norm_counts", "ps_vst", "vsd"), 
     file = "./processed_physeq.rda")
```


```{r, eval = FALSE}
save(list = c("ps", "pscontrols", "ps_asinh", "SMP", "SUBJ", "dds", "norm_counts"), 
     file = "./processed_physeq.rda")
```



Data Visualization {#dat-vis .unnumbered} 
=========================================

```{r}
taxtab <- data.frame(tax_table(ps)) %>%
  mutate(Seq_ID = rownames((.)))
```


PCA {#pca .unnumbered}
---------------------------------

```{r}
load("./processed_physeq.rda")
load("ordinate_16S.rda")
```


```{r pca, eval = FALSE}
# Compute PCA 
pca.ihs <- prcomp(t(asinh(norm_counts)), scale = FALSE)
save(list = c("pca.ihs"), file = "./ordinate_16S.rda")
```


```{r}
library("factoextra")
library("ggrepel")
fviz_eig(pca.ihs)
```

```{r pca-subj, fig.height=10, fig.width=14}
fviz_pca_ind(pca.ihs, habillage= SMP$Subject, 
             label = "none", geom = "point", pointshape = 16, 
             addEllipses = TRUE, ellipse.level = 0.6) + 
  coord_fixed(pca.ihs$sdev[2]/pca.ihs$sdev[1])
```


```{r}
nPC <- 5
loadings <- pca.ihs$rotation[, 1:nPC] %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  rownames_to_column("Seq_ID") %>%
  left_join(taxtab)

scores <- pca.ihs$x[, 1:nPC] %>%
  as.data.frame() %>%
  rownames_to_column("Meas_ID") %>%
  left_join(SMP) %>%
  arrange(Group, Subject, Samp_Date)  %>%
  mutate(constant = "constant")
```


```{r pca-intv, fig.height=7, fig.width=14}
cols <- c("grey77", colorRampPalette(brewer.pal(9, "Set1"))(12))
names(cols) <- c("NoInterv", setdiff(unique(scores$Interval), "NoInterv"))
plot_projection(
  scores, xname = "PC1", yname = "PC2", labname = NULL,
  size = 2, color = "Interval", eigs = pca.ihs$sdev^2) +
  scale_color_manual(values = cols) +
  theme(legend.position = "right")
```

```{r pca-intv-2-3, fig.height=9, fig.width=12}
plot_projection(
  scores, xname = "PC3", yname = "PC4", labname = NULL,
  size = 2, color = "Interval", eigs = pca.ihs$sdev^2) +
  scale_color_manual(values = cols) +
  theme(legend.position = "right")
```

```{r pca-loadings, fig.height=10, fig.width=15}
plt <- plot_projection(
  loadings, xname = "PC1", yname = "PC2", size = "PC3",
  color = "Class", eigs = pca.ihs$sdev^2, alpha = 0.5
  ) +
  geom_text_repel(
    data = loadings %>% 
      filter((PC1^2 + PC2^2) > 0.006), 
    aes(label = Genus),
    nudge_y = 0.002) +
  theme(legend.position = "right")
print(plt)
```


```{r pca-subj-time, fig.height=40, fig.width=20, warning = FALSE}
plot_projection(scores, xname = "PC1", yname = "PC2", labname = NULL, 
                size = 5, color = "constant", eigs = pca.ihs$sdev^2) +
  geom_point(data = scores %>% 
               filter(Interval != "NoInterv"), aes(fill = Interval),
             size = 5, pch = 23, color = "black", lwd=2) +
  facet_wrap(Group ~ Subject, scales = "free", ncol = 5) +
  scale_color_manual(values  = "grey77") +
  scale_fill_manual(values = colorRampPalette(brewer.pal(9, "Set1"))(20))
```

Centering the subjects

```{r}
subject_scores <- scores %>%
  select(Subject, PC1:PC5) %>%
  group_by(Subject) %>%
  summarise_all(mean)

scores_centered <- scores %>%
  left_join(subject_scores, by = c("Subject"), suffix = c("", ".subj")) %>%
  mutate(
    PC1 = PC1 - PC1.subj,
    PC2 = PC2 - PC2.subj,
    PC3 = PC3 - PC3.subj,
    PC4 = PC4 - PC4.subj,
    PC5 = PC5 - PC5.subj
  )
```

```{r pca-intv-centered-abx, fig.height=6, fig.width=14}
plot_projection(
  scores_centered, xname = "PC1", yname = "PC2", labname = NULL,
  size = 2.5, color = "Abx_Interval", eigs = pca.ihs$sdev^2, alpha = 0.5) +
  geom_point(data = scores_centered %>% 
                filter(Interval == "MidAbx"), aes(color = Interval),
             size = 2.5) +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "right") +
  facet_wrap(~Group)
```

```{r pca-intv-centered-diet, fig.height=6, fig.width=14}
plot_projection(
  scores_centered, xname = "PC1", yname = "PC2", labname = NULL,
  size = 2, color = "Diet_Interval", eigs = pca.ihs$sdev^2, alpha = 0.5) +
  geom_point(data = scores_centered %>% 
                filter(Interval == "MidDiet"), aes(color = Interval),
             size = 2.5) +
  scale_color_brewer(palette = "Set1") +
  theme(legend.position = "right") +
  facet_wrap(~Group)
```

```{r pca-intv-centered-cc, fig.height=6, fig.width=14}
plot_projection(
  scores_centered, xname = "PC1", yname = "PC2", labname = NULL,
  size = 2, color = "CC_Interval", eigs = pca.ihs$sdev^2, alpha = 0.5) +
  scale_color_brewer(palette = "Dark2") +
  theme(legend.position = "right") +
  facet_wrap(~Group)
```


```{r pca-intv-centered-time, fig.height=6, fig.width=14}
plot_projection(
  scores_centered, xname = "PC1", yname = "PC2", labname = NULL,
  size = 2, color = "DaysFromStart", eigs = pca.ihs$sdev^2, alpha = 0.5) +
  scale_color_viridis(trans = "sqrt") +
  theme(legend.position = "right") +
  facet_wrap(~Group)
```


Sparse PCA {#pcoa .unnumbered}
------------------------------

```{r}
sparse_pca <- PMA::SPC(
  scale(t(asinh(norm_counts)), center = TRUE, scale = FALSE), 
  K = 5, sumabsv = 30)
save(list = c("pca.ihs", "sparse_pca"), file = "./ordinate_16S.rda")

```



```{r}
nPC <- 5
loadings <- 
  left_join(taxtab)

scores <- pca.ihs$x[, 1:nPC] %>%
  as.data.frame() %>%
  rownames_to_column("Meas_ID") %>%
  left_join(SMP) %>%
  arrange(Group, Subject, Samp_Date)  %>%
  mutate(constant = "constant")
```

adaptiveGPCA {#agPCA .unnumbered}
---------------------------------

```{r, echo = FALSE, eval = FALSE}
library("adaptiveGPCA")
pp <- processPhyloseq(ps)
out.agpca <- adaptivegpca(pp$X, pp$Q, k = 5)

pp.ihs <- processPhyloseq(psIHS)
out.agpca.ihs <- adaptivegpca(pp.ihs$X, pp.ihs$Q, k = 5)

# out.ff = gpcaFullFamily(pp$X, pp$Q, k = 2)
# out.agpca <- 
#   visualizeFullFamily(out.ff,
#                       sample_data = sample_data(ps),
#                       sample_mapping = aes(x = Axis1, y = Axis2, color = type),
#                       var_data = tax_table(ps),
#                       var_mapping = aes(x = Axis1, y = Axis2, color = Class))

save(list = c("bray.pcoa.raw", "bray.pcoa.ihs", "brayD", "pca.ihs", "bray.rtsne.ihs",
            "out.agpca", "out.agpca.ihs"), 
   file = "../data/processed/perturb_ordinate.rda")
            
```

```{r, eval=FALSE}
sample.scores <- cbind(agPCA1 = out.agpca.ihs$U[, 1],
                       agPCA2 = out.agpca.ihs$U[, 2],
                       agPCA3 = out.agpca.ihs$U[, 3],
                       sample.scores) 
```

```{r agpca-subj, echo = FALSE, eval= FALSE, fig.height=10, fig.width=14}
plot_projection(sample.scores, xname = "agPCA1", yname = "agPCA2", 
            size = 3, color = "Subject", eigs = out.agpca.ihs$vars)
```

```{r, eval=FAlSE}
loadings <- data.frame(out.agpca.ihs$QV, tax_table(psIHS)) %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  rownames_to_column("Seq_ID") %>%
  left_join(taxtab)
```


```{r agpca-loadings,  eval=FAlSE, fig.height=15, fig.width=20}
plot_projection(loadings, 
                xname = "Axis1", yname = "Axis2", size = "Axis3",
                color = "genus", eigs = out.agpca.ihs$vars, alpha = 0.3) +
  geom_text_repel(
    data = loadings %>% 
      filter(abs(Axis1) + abs(Axis2) > 0.035), 
    aes(label = species), 
    force = 0.002) + 
  theme(legend.position = "bottom")

plot_projection(loadings, 
                xname = "Axis1", yname = "Axis2", size = "Axis3",
                color = "Class", eigs = out.agpca.ihs$vars, alpha = 0.3) +
  geom_text_repel(
    data = loadings %>% 
      filter(abs(Axis1) + abs(Axis2) > 0.035), 
    aes(label = genus), 
    force = 0.002) + 
  theme(legend.position = "bottom")
```


```{r agpca-time,  eval=FAlSE, fig.height=30, fig.width=18}
theme_set(theme_bw())
plot_projection(sample.scores, 
                xname = "agPCA1", yname = "agPCA2", labname = NULL, 
                size = 3, color = "DayFromStart",
                eigs = out.agpca.ihs$vars) +
  geom_point(data = sample.scores %>% 
               filter(Timeline != "typical"), aes(fill = Timeline),
             size = 5, pch = 23, color = "black", lwd=2) +
  facet_wrap(Group~Subject, scales = "free", ncol = 5) +
  scale_color_viridis() +
  scale_fill_brewer(palette = "Oranges")
```


## TSNE {#tsne .unnumbered}

```{r, eval = FALSE}
set.seed(123)
bray.rtsne.ihs <- Rtsne::Rtsne(brayD.ihs, is_distance = TRUE, dims = 2, 
                               pca = TRUE,
                               eta = 1, exaggeration_factor = nsamples(ps)/10)
save(list = c("bray.pcoa.raw", "bray.pcoa.ihs", "brayD", "pca.ihs", "bray.rtsne.ihs"), 
     file = "../data/processed/perturb_ordinate.rda")
```

```{r}
bray.rtsne.ihs <- Rtsne::Rtsne(brayD.ihs, is_distance = TRUE, dims = 2, 
                               pca = TRUE)
sample.scores2 <- cbind(tsne.1.ihs = bray.rtsne.ihs$Y[, 1],
                        tsne.2.ihs = bray.rtsne.ihs$Y[, 2],
                        sample.scores) 
```


```{r}
sample.scores <- cbind(tsne.1.ihs = bray.rtsne.ihs$Y[, 1],
                       tsne.2.ihs = bray.rtsne.ihs$Y[, 2],
                       sample.scores) 
```



```{r tsne-bray-subj, fig.height=10, fig.width=14}
plot_projection(sample.scores, xname = "tsne.1.ihs", yname = "tsne.2.ihs", 
            size = 3, color = "Subject")

plot_projection(sample.score2s, xname = "tsne.1.ihs", yname = "tsne.2.ihs", 
            size = 3, color = "Subject")
```

```{r tsne-bray-time, fig.height=10, fig.width=14}
plot_projection(sample.scores, 
                xname = "tsne.1.ihs", yname = "tsne.2.ihs", labname = NULL,
                size = 3, color = "DayFromStart") + 
  scale_color_viridis()

plot_projection(sample.scores2, 
                xname = "tsne.1.ihs", yname = "tsne.2.ihs", labname = NULL,
                size = 3, color = "DayFromStart") + 
  scale_color_viridis()
```


```{r tsne-bray-subj-time,  fig.height=30, fig.width=18}

plot_projection(sample.scores %>%
                  arrange(Subject, Samp_Date), 
                xname = "tsne.1.ihs", yname = "tsne.2.ihs", labname = NULL, 
            size = 3, color = "DayFromStart") +
  geom_path(
        aes(group = Subject, color = DayFromStart), 
        alpha = 0.8, lwd = 1) +
  geom_point(data = sample.scores %>% 
               filter(Timeline != "typical"), aes(fill = Timeline),
             size = 5, pch = 23, color = "black", lwd=2) +
  facet_wrap(Group~Subject, scales = "free", ncol = 5) +
  scale_color_viridis() +
  scale_fill_brewer(palette = "Oranges")
```






```{r}

```






```{r}
sessionInfo()
```

