---
title: "Resilience Study 16S data: tree Discriminant Analysis"
author: 
  - name: Lan Huong Nguyen
    affiliation: Institute for Computational and Mathematical Engineering, Stanford University, CA 94305
keywords: microbiome, metatranscriptomics.
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{Gene expression time course data analysis}
  %\VignetteEngine{knitr::rmarkdown}
---


```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("rmarkdown")
options(width = 200, stringsAsFactors = FALSE) 
knitr::opts_chunk$set(
  message = FALSE, error = FALSE, warning = FALSE, 
  fig.width = 8, fig.height = 6,
  fig.path = "../figs/holmes_group_presentation/treeDA/", 
  dev='png') 
```

```{r warning=FALSE, message=FALSE}
#rm(list = ls())
library("DESeq2")
library("viridis")
library("RColorBrewer")
library("ggplot2")
library("tidyverse")
library("phyloseq")
library("treeDA")
library("adaptiveGPCA")

datadir <- "/scratch/PI/sph/resilience/"
curdir <- getwd()
resfile <- "results/treeDA_res16S.rda"
source('../utils.R')
saveRES <- TRUE
```

```{r}
theme_set(theme_bw())
theme_update(
   panel.border = element_rect(size = 0.5),
   legend.position = "bottom",
   text = element_text(size = 15),
   strip.background = element_blank()
)
```

```{r}
processed_datafile <- "results/processed_physeq.rda"
load(processed_datafile)
#load("results/ordinate_16S.rda")
```

```{r}
dds <- phyloseq_to_deseq2(ps, design = ~ 1)
dds <- estimateSizeFactors(dds, type = "poscounts")
norm_counts <- counts(dds, normalized = TRUE)
ps_norm <- ps
otu_table(ps_norm) <- otu_table(norm_counts, taxa_are_rows = TRUE)
ps_norm
```


# Tree Discriminant Analysis

## Response to Diet

```{r}
# Subest to only diet pre/post diet but before other perturbations
diet <- subset_samples(ps_norm, grepl("Diet", Group))
diet <- subset_samples(
  diet, !grepl("MidAbx", Abx_Interval) & !grepl("PostAbx", Abx_Interval) &
    !grepl("PostCC", CC_Interval))
diet

# Filter taxa that occurs at minTaxaSum in at least minNoSubj subjects
minTaxaSum <- 10; minTaxaPrev <- 3; minNoSubj <- 3
ASVsum <- data.frame(t(as(otu_table(diet), "matrix"))) %>%
  mutate(Subject = diet@sam_data$Subject) %>%
  group_by(Subject) %>%
  summarise_all(sum)
ASVprev <- data.frame(t(as(otu_table(diet), "matrix")) > 0) %>%
  mutate(Subject = diet@sam_data$Subject) %>%
  group_by(Subject) %>%
  summarise_all(sum)

keepASV <- data.frame(
  Seq_ID = colnames(ASVsum),
  abundant = colSums(ASVsum >= minTaxaSum) >= minNoSubj, 
  prevalent = colSums(ASVprev >= minTaxaPrev) >= minNoSubj) %>%
  filter(abundant, prevalent)

diet <- prune_taxa(keepASV$Seq_ID, diet)
diet
diet_SMP <- data.frame(diet@sam_data) %>%
  mutate(Samp_Date = as.Date(Samp_Date))
```


```{r exp-design, fig.width=8, fig.height=8}
new_intv_levs <- unique(diet_SMP$Interval)
colors <- colorRampPalette(brewer.pal(8, "Dark2"))(length(new_intv_levs)-2)
colors <- c("grey57", "red", colors)
names(colors) <- c("NoInterv", "MidDiet", "PostDiet_PreAbx", "PostDiet_PreCC",
                   "PostDiet_PreCC_PreAbx")
ggplot(
  diet_SMP, aes(y = Subject, x = Diet_RelDay)) +
  geom_point(
      aes(color = Interval, fill = Interval), 
      pch = 124, size = 3) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    theme(legend.direction = "horizontal", legend.box = "vertical",
          strip.text.y = element_text(angle=270, size = 10)) +
    guides(color = guide_legend(nrow = 2 , override.aes = list(size=3, pch = 16)))
```

```{r}
sample_data(diet)$type <- with(sample_data(diet), 
  memisc::cases(
    Diet_RelDay < -15                    -> "Ancient",
    Diet_RelDay < 0 & Diet_RelDay >= -15 -> "PreDiet",
    Diet_RelDay < 7 & Diet_RelDay >= 0   -> "MidDiet",
    Diet_RelDay < 20 & Diet_RelDay >= 7  -> "PostDiet",
    Diet_RelDay >= 20                    -> "Recovery"
  )
)
```

```{r}
table(sample_data(diet)$type)
```

### Two classes only

```{r}
diet_change <- subset_samples(diet, type %in%  c("PostDiet", "PreDiet")) 
diet_change <- subset_taxa(diet_change, taxa_sums(diet_change) > 0)
diet_change
```
```{r}
table(sample_data(diet_change)$type)
```


```{r}
diet.treeda = treeda(
  response = sample_data(diet_change)$type,
  predictors = asinh(t(otu_table(diet_change))),
  tree = phy_tree(diet_change), 
  p = 100)
```

```{r}
diet.treeda
```

```{r}
ggplot(data.frame(sample_data(diet_change), 
                  projections = diet.treeda$projections)) +
  geom_point(aes(x = Subject, y = projections, color = type)) +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
nTaxa <- 20
beta <- as.numeric(diet.treeda$leafCoefficients$beta)
taxtab <- data.frame(
  beta = beta,
  as(tax_table(diet_change), "matrix"), 
  stringsAsFactors = FALSE)
freq_family <- taxtab %>%
  group_by(Family) %>%
  summarise(Freq = n()) %>%
  filter(!is.na(Family)) %>%
  arrange(desc(Freq))

family_levels <- unique(c(
  freq_family$Family[1:nTaxa]),
  taxtab$Family[abs(beta) > 1e-3])[1:nTaxa]

taxtab$family <- factor(taxtab$Family, levels = family_levels)
```

```{r fig.width=12, fig.height=8}
plts <- plot_coefficients(diet.treeda, color = taxtab$family)
combine_plot_and_tree(plts$coef.plot, plts$tree.plot, tree.height = 2)
```


```{r fig.width=12, fig.height=8}
df <- plts$df %>% 
      mutate(
        Family = taxtab$family,
        Species = ifelse(is.na(taxtab$Species),"", taxtab$Species),
        Genus = ifelse(is.na(taxtab$Genus), "", taxtab$Genus),
        org.name = paste(Genus, Species),
        predictor.names = ifelse(org.name == " ", predictor.names, org.name)) %>%
  filter(abs(value) > 0.009)

df <- df[!duplicated(df[, c("value", "predictor.names")]),]

coef.plot <- plts$coef.plot +
  ggrepel::geom_text_repel(
    data = df,
    aes(x = leaf.position, y = value, label = predictor.names, color = Family)
  )
coef.plot
```




```{r}
thresh <- 1e-3
non_zero_taxa <- taxtab %>% 
  rownames_to_column("Seq_ID") %>%
  select(-Seq) %>% 
  filter(abs(beta) > thresh) %>%
  arrange(desc(abs(beta)))

DF <- data.frame(asinh(as(otu_table(diet), "matrix"))) %>%
  rownames_to_column("Seq_ID") %>%
  filter(Seq_ID %in% non_zero_taxa$Seq_ID) %>%
  gather(Meas_ID, abundance, -Seq_ID) %>%
  left_join(
    non_zero_taxa %>%
      select(Seq_ID:Species, family)) %>%
  left_join(diet_SMP) %>%
  mutate(
    Diet_Interval = factor(
      Diet_Interval, levels = c("PreDiet", "MidDiet", "PostDiet")))

DF_reference <- DF %>%
  filter(Diet_RelDay >= -7, Diet_RelDay < 0) %>%
  group_by(Seq_ID, Subject) %>%
  summarise(
    reference_abundance = mean(abundance),
    Freq = n(),
    subj_seq_abund_sd = sqrt(sum((abundance-reference_abundance)^2)/(Freq-1)),
    subj_seq_abund_sd = ifelse(subj_seq_abund_sd == 0, 1, subj_seq_abund_sd)
  )

DF <- DF %>%
  left_join(DF_reference) %>%
  mutate(
    abundance_centered = abundance - reference_abundance,
    abundance_centered_scaled = abundance_centered/subj_seq_abund_sd)

```

```{r}
freq_genus <- non_zero_taxa %>%
  filter(!is.na(Genus)) %>%
  group_by(Genus) %>%
  summarise(count = n()) %>%
  arrange(desc(count))
freq_genus
```

```{r fig.height=10, fig.width=10}
ggplot(
  DF %>% filter(Genus %in% "Clostridium_IV") %>%
    filter(abundance > 0) %>%
    arrange(Subject, Diet_RelDay),
  aes(x = abundance)) +
  geom_violin(aes(x = Diet_Interval, y = abundance)) +
  # geom_histogram(aes(x = abundance, fill = Diet_Interval), position = "dodge") +
  geom_jitter(
    aes(x = Diet_Interval, y = abundance, color = Subject),
    height = 0, width = 0.15) +
  facet_wrap(~ Seq_ID, scales = "free") +
  coord_flip()
```


```{r, fig.height=10, fig.width=10}
ggplot(
  DF %>% 
    filter(Genus %in% "Clostridium_IV") %>%
    arrange(Subject, Diet_RelDay),
  aes(y = abundance, x = Diet_RelDay)) +
  geom_line(aes(group = Subject)) +
  geom_smooth() +
  facet_wrap(~ Seq_ID, scales = "free") 
```


## Multiple classes

```{r}
# diet.treeda.mult = treeda(
#   response = sample_data(diet)$type,
#   predictors = asinh(t(otu_table(diet))),
#   tree = phy_tree(diet), 
#   p = 50)
# diet.treeda.mult

```


```{r}
# ggplot(data.frame(sample_data(diet), 
#                   projections = diet.treeda.mult$projections)) +
#   geom_point(aes(x = Subject, y = projections, color = type)) +
#   theme(axis.text.x = element_text(angle = 90))
```


```{r}
# nTaxa <- 20
# beta <- as.numeric(diet.treeda.mult$leafCoefficients$beta)
# freq_family <- taxtab %>%
#   group_by(Family) %>%
#   summarise(Freq = n()) %>%
#   filter(!is.na(Family)) %>%
#   arrange(desc(Freq))
# 
# family_levels <- unique(c(
#   freq_family$Family[1:nTaxa]),
#   unique(taxtab$Family[abs(beta) > 1e-3]))[1:nTaxa]
# 
# taxtab <- data.frame(as(tax_table(diet), "matrix"), stringsAsFactors = FALSE)
# taxtab$family <- factor(taxtab$Family, levels = family_levels)
```


```{r}
# plot_coefficients(diet.treeda.mult, color = taxtab$family)
```


## Response to Antibiotics

```{r}
abx <- subset_samples(ps_norm, grepl("Abx", Group))
abx <- subset_samples(abx, 
  !grepl("MidDiet", Diet_Interval) & !grepl("PreDiet", Diet_Interval) &
    !grepl("PreCC", CC_Interval))

# Filter taxa that occurs at minTaxaSum in at least minNoSubj subjects
minTaxaSum <- 10; minTaxaPrev <- 3; minNoSubj <- 3
ASVsum <- data.frame(t(as(otu_table(abx), "matrix"))) %>%
  mutate(Subject = abx@sam_data$Subject) %>%
  group_by(Subject) %>%
  summarise_all(sum)
ASVprev <- data.frame(t(as(otu_table(abx), "matrix")) > 0) %>%
  mutate(Subject = abx@sam_data$Subject) %>%
  group_by(Subject) %>%
  summarise_all(sum)

keepASV <- data.frame(
  Seq_ID = colnames(ASVsum),
  abundant = colSums(ASVsum >= minTaxaSum) >= minNoSubj, 
  prevalent = colSums(ASVprev >= minTaxaPrev) >= minNoSubj) %>%
  filter(abundant, prevalent)

abx <- prune_taxa(keepASV$Seq_ID, abx)
abx
abx_SMP <- data.frame(abx@sam_data) %>%
  mutate(Samp_Date = as.Date(Samp_Date))
```

```{r exp-design, fig.width=8, fig.height=8}
new_intv_levs <- unique(abx_SMP$Interval)
colors <- colorRampPalette(brewer.pal(8, "Dark2"))(length(new_intv_levs)-2)
colors <- c("grey57", "red", colors)
names(colors) <- c("NoInterv", "MidAbx", "PostDiet_PreAbx", 
                   "PostDiet_PostAbx", "PostDiet_PostCC_PreAbx", 
                   "PostDiet_PostCC_PostAbx", "PostAbx", "UnpAbx")
ggplot(
  abx_SMP, aes(y = Subject, x = Abx_RelDay)) +
  geom_point(
      aes(color = Interval, fill = Interval), 
      pch = 124, size = 3) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    theme(legend.direction = "horizontal", legend.box = "vertical",
          strip.text.y = element_text(angle=270, size = 10)) +
    guides(color = guide_legend(nrow = 3 , override.aes = list(size=3, pch = 16)))
```



```{r}
sample_data(abx)$type <- with(sample_data(abx), 
  memisc::cases(
    Abx_RelDay < -20                   -> "Ancient",
    Abx_RelDay < 0 & Abx_RelDay >= -20 -> "PreAbx",
    Abx_RelDay < 7 & Abx_RelDay >= 0   -> "MidAbx",
    Abx_RelDay < 30 & Abx_RelDay >= 7  -> "PostAbx",
    Abx_RelDay >= 30                   -> "Recovery"
  )
)
```

```{r}
table(sample_data(abx)$type)
```


### Two classes only

```{r}
abx_change <- subset_samples(abx, type %in%  c("PostAbx", "PreAbx")) 
abx_change <- subset_taxa(abx_change, taxa_sums(abx_change) > 0)
abx_change
```

```{r}
table(sample_data(abx_change)$type)
```

```{r}
abx.treeda = treeda(
  response = sample_data(abx_change)$type,
  predictors = asinh(t(otu_table(abx_change))),
  tree = phy_tree(abx_change), 
  p = 50)
```

```{r}
abx.treeda
```

```{r}
ggplot(data.frame(sample_data(abx_change), 
                  projections = abx.treeda$projections)) +
  geom_point(aes(x = Subject, y = projections, color = type)) +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
nTaxa <- 20
beta <- as.numeric(abx.treeda$leafCoefficients$beta)
taxtab <- data.frame(
  beta = beta,
  as(tax_table(abx_change), "matrix"), 
  stringsAsFactors = FALSE)

freq_family <- taxtab %>%
  group_by(Family) %>%
  summarise(Freq = n()) %>%
  filter(!is.na(Family)) %>%
  arrange(desc(Freq))

family_levels <- unique(c(
  freq_family$Family[1:nTaxa]),
  taxtab$Family[abs(beta) > 1e-3])[1:nTaxa]

taxtab$family <- factor(taxtab$Family, levels = family_levels)
```


```{r fig.width=12, fig.height=8}
plts <- plot_coefficients(abx.treeda, color = taxtab$family)
combine_plot_and_tree(plts$coef.plot, plts$tree.plot, tree.height = 2)
```


```{r fig.width=12, fig.height=8}
df <- plts$df %>% 
      mutate(
        Family = taxtab$family,
        Species = ifelse(is.na(taxtab$Species),"", taxtab$Species),
        Genus = ifelse(is.na(taxtab$Genus), "", taxtab$Genus),
        org.name = paste(Genus, Species),
        predictor.names = ifelse(org.name == " ", predictor.names, org.name)) %>%
  filter(abs(value) > 0.002)

df <- df[!duplicated(df[, c("value", "predictor.names")]),]

coef.plot <- plts$coef.plot +
  ggrepel::geom_text_repel(
    data = df,
    aes(x = leaf.position, y = value, label = predictor.names, color = Family)
  )
coef.plot
```




```{r}
thresh <- 1e-3
non_zero_taxa <- taxtab %>% 
  rownames_to_column("Seq_ID") %>%
  select(-Seq) %>% 
  filter(abs(beta) > thresh) %>%
  arrange(desc(abs(beta)))
X <- t(asinh(as(otu_table(abx), "matrix")))
X <- sweep(X, 2, STATS = colMeans(X), FUN = "-")
sds <- apply(X, 2, sd)
sds[sds == 0] <- 1
X <- sweep(X, 2, STATS = sds, FUN = "/")

DF <- data.frame(t(X)) %>%
  rownames_to_column("Seq_ID") %>%
  filter(Seq_ID %in% non_zero_taxa$Seq_ID) %>%
  gather(Meas_ID, abundance, -Seq_ID) %>%
  left_join(
    non_zero_taxa %>%
      select(Seq_ID:Species, family)) %>%
  left_join(abx_SMP) %>%
  mutate(
    Abx_Interval = factor(
      Abx_Interval, levels = c("PreAbx", "MidAbx", "PostAbx")))

DF_reference <- DF %>%
  filter(Abx_RelDay >= -7, Abx_RelDay < 0) %>%
  group_by(Seq_ID, Subject) %>%
  summarise(
    reference_abundance = mean(abundance),
    Freq = n(),
    subj_seq_abund_sd = sqrt(sum((abundance-reference_abundance)^2)/(Freq-1)),
    subj_seq_abund_sd = ifelse(subj_seq_abund_sd == 0, 1, subj_seq_abund_sd)
  )

DF <- DF %>%
  left_join(DF_reference) %>%
  mutate(
    abundance_centered = abundance - reference_abundance,
    abundance_centered_scaled = abundance_centered/subj_seq_abund_sd)

```

```{r}
freq_genus <- non_zero_taxa %>%
  filter(!is.na(Genus)) %>%
  group_by(Genus) %>%
  summarise(count = n()) %>%
  arrange(desc(count))
freq_genus
```

```{r fig.height=10, fig.width=10}
ggplot(
  DF %>% filter(Genus %in% "Blautia") %>%
    #filter(abundance > 0) %>%
    arrange(Subject, Abx_RelDay),
  aes(x = abundance)) +
  geom_boxplot(aes(x = Abx_Interval, y = abundance)) +
  #geom_violin(aes(x = Abx_Interval, y = abundance)) +
  # geom_jitter(
  #   aes(x = Abx_Interval, y = abundance, color = Subject),
  #   height = 0, width = 0.15) +
  facet_wrap(~ Seq_ID, scales = "free") +
  coord_flip()
```


```{r, fig.height=10, fig.width=10}
ggplot(
  DF %>% 
    filter(Genus %in% "Clostridium_XVIII") %>%
    arrange(Subject, Abx_RelDay),
  aes(y = abundance_centered, x = Abx_RelDay)) +
  geom_point(aes(color = Subject)) +
  geom_smooth() +
  facet_wrap(~ Seq_ID, scales = "free") 
```

### Two classes only Pre vs Mid

```{r}
abx_change <- subset_samples(
  abx, (type == "MidAbx" & Abx_RelDay >= 1) | (type == "PreAbx" & Abx_RelDay >= -10)) 

abx_change <- subset_taxa(abx_change, taxa_sums(abx_change) > 0)
abx_change
```

```{r}
table(sample_data(abx_change)$type)
```

```{r}
abx.treeda = treeda(
  response = sample_data(abx_change)$type,
  predictors = asinh(t(otu_table(abx_change))),
  tree = phy_tree(abx_change), 
  p = 50)
```

```{r}
abx.treeda
```

```{r}
ggplot(data.frame(sample_data(abx_change), 
                  projections = abx.treeda$projections)) +
  geom_point(aes(x = Subject, y = projections, color = type)) +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
nTaxa <- 20
beta <- as.numeric(abx.treeda$leafCoefficients$beta)
taxtab <- data.frame(
  beta = beta,
  as(tax_table(abx_change), "matrix"), 
  stringsAsFactors = FALSE)
freq_family <- taxtab %>%
  group_by(Family) %>%
  summarise(Freq = n()) %>%
  filter(!is.na(Family)) %>%
  arrange(desc(Freq))

family_levels <- unique(c(
  freq_family$Family[1:nTaxa]),
  taxtab$Family[abs(beta) > 1e-3])[1:nTaxa]

taxtab$family <- factor(taxtab$Family, levels = family_levels)
```


```{r fig.width=12, fig.height=8}
plts <- plot_coefficients(abx.treeda, color = taxtab$family)
combine_plot_and_tree(plts$coef.plot, plts$tree.plot, tree.height = 2)
```


```{r fig.width=12, fig.height=8}
df <- plts$df %>% 
      mutate(
        Family = taxtab$family,
        Species = ifelse(is.na(taxtab$Species),"", taxtab$Species),
        Genus = ifelse(is.na(taxtab$Genus), "", taxtab$Genus),
        org.name = paste(Genus, Species),
        predictor.names = ifelse(org.name == " ", predictor.names, org.name)) %>%
  filter(abs(value) > 0.005)

df <- df[!duplicated(df[, c("value", "predictor.names")]),]

coef.plot <- plts$coef.plot +
  ggrepel::geom_text_repel(
    data = df,
    aes(x = leaf.position, y = value, label = predictor.names, color = Family)
  )
coef.plot
```



```{r}
thresh <- 1e-3
non_zero_taxa <- taxtab %>% 
  rownames_to_column("Seq_ID") %>%
  select(-Seq) %>% 
  filter(abs(beta) > thresh) %>%
  arrange(desc(abs(beta)))
X <- t(asinh(as(otu_table(abx), "matrix")))
X <- sweep(X, 2, STATS = colMeans(X), FUN = "-")
sds <- apply(X, 2, sd)
sds[sds == 0] <- 1
X <- sweep(X, 2, STATS = sds, FUN = "/")

DF <- data.frame(t(X)) %>%
  rownames_to_column("Seq_ID") %>%
  filter(Seq_ID %in% non_zero_taxa$Seq_ID) %>%
  gather(Meas_ID, abundance, -Seq_ID) %>%
  left_join(
    non_zero_taxa %>%
      select(Seq_ID:Species, family)) %>%
  left_join(abx_SMP) %>%
  mutate(
    Abx_Interval = factor(
      Abx_Interval, levels = c("PreAbx", "MidAbx", "PostAbx")))

DF_reference <- DF %>%
  filter(Abx_RelDay >= -7, Abx_RelDay < 0) %>%
  group_by(Seq_ID, Subject) %>%
  summarise(
    reference_abundance = mean(abundance),
    Freq = n(),
    subj_seq_abund_sd = sqrt(sum((abundance-reference_abundance)^2)/(Freq-1)),
    subj_seq_abund_sd = ifelse(subj_seq_abund_sd == 0, 1, subj_seq_abund_sd)
  )

DF <- DF %>%
  left_join(DF_reference) %>%
  mutate(
    abundance_centered = abundance - reference_abundance,
    abundance_centered_scaled = abundance_centered/subj_seq_abund_sd)

```

```{r}
freq_genus <- non_zero_taxa %>%
  filter(!is.na(Genus)) %>%
  group_by(Genus) %>%
  summarise(count = n()) %>%
  arrange(desc(count))
freq_genus
```

### Multiple Classes


```{r}
abx.treeda.mult = treeda(
  response = sample_data(abx)$type,
  predictors = t(otu_table(abx)),
  tree = phy_tree(abx), p = 50)
```

```{r}
nTaxa <- 10
beta <- as.numeric(abx.treeda.mult$leafCoefficients$beta)
taxtab <- data.frame(as(tax_table(abx), "matrix"), stringsAsFactors = FALSE)
freq_family <- taxtab %>%
  filter(!is.na(Family)) %>%
  group_by(Family) %>%
  summarise(Freq = n()) %>%
  arrange(desc(Freq))
taxa_levels <- unique(
  c(freq_family$Family[1:nTaxa], taxtab$Family[abs(beta) > 1e-6]))
taxtab$family <- factor(taxtab$Family, levels = taxa_levels)
```


```{r}
abx.treeda.mult
```

```{r}
ggplot(data.frame(sample_data(abx), projections = abx.treeda.mult$projections)) +
  geom_point(aes(x = Subject, y = projections.1, color = type)) +
  theme(axis.text.x = element_text(angle = 90))
```

```{r}
ggplot(data.frame(sample_data(abx), projections = abx.treeda.mult$projections)) +
  geom_point(aes(x = Subject, y = projections.2, color = type)) +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
ggplot(data.frame(sample_data(abx), projections = abx.treeda.mult$projections)) +
  geom_point(aes(x = Subject, y = projections.3, color = type)) +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
plot_coefficients(abx.treeda.mult, color = rep(taxtab$family, 3))
```




```{r}

```






```{r}
sessionInfo()
```

