---
title: "Presentation Plots"
author: 
  - name: Lan Huong Nguyen
    affiliation: Institute for Computational and Mathematical Engineering, Stanford University, CA 94305
keywords: microbiome, metatranscriptomics.
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document2:
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{Gene expression time course data analysis}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
rm(list = ls())
library("BiocStyle")
library("rmarkdown")
options(width = 200) 
knitr::opts_chunk$set(
  message = FALSE, error = FALSE, warning = FALSE, 
  fig.width = 8, fig.height = 6,
  fig.path = "figs/", dev='png') 
```

```{r setup}
library("tidyverse")
library("readxl")
library("data.table")
library("gridExtra")

theme_set(theme_bw())
theme_updates <- theme(text = element_text(size =20))
```


Read Coverage {#read-coverage .unnumbered}
========================================

Here we summarize the read coverage statistics for all data modalities.

16S {#read-coverage-16S .unnumbered}
------------------------------------

```{r}
datadir <- "/scratch/PI/sph/resilience/"
physeqfile <- file.path(datadir, "16S/phyloseq/perturb_physeq_22Jan18.rds")
stats16sfile <- file.path(datadir, "16S/16S_dada2_results/track_all_pools.csv")
ps <- readRDS(physeqfile)
ps
stats16s <-read.csv(stats16sfile, row.names = 1)
head(stats16s)
```





MetaGenomics {#read-coverage-metag .unnumbered}
------------------------------------

```{r}

```

MetaTranscriptomics {#read-coverage-metat .unnumbered}
------------------------------------

```{r}

```


```{r}
sessionInfo()
```

