---
title: "16S Data Analysis"
author: 
  - name: Lan Huong Nguyen
    affiliation: Institute for Computational and Mathematical Engineering, Stanford University, CA 94305
keywords: microbiome, metatranscriptomics.
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{Gene expression time course data analysis}
  %\VignetteEngine{knitr::rmarkdown}
---

Edited July 15th, 2019


```{r, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
library("BiocStyle")
library("rmarkdown")
options(width = 200, stringsAsFactors = FALSE) 
knitr::opts_chunk$set(
  message = FALSE, error = FALSE, warning = FALSE, 
  fig.width = 8, fig.height = 6,
  fig.path = "../figs/analysis_summer/generate_phylseq/", 
  dev='png') 
```

```{r setup, warning=FALSE, message=FALSE}
pkgs_needed <- c("phyloseq", "tidyverse", "RColorBrewer", "DESeq2",
               "data.table", "readxl", "BiocManager", "decontam")
BiocManager::install(setdiff(pkgs_needed, installed.packages()))

library("RColorBrewer")
library("phyloseq")
library("tidyverse")

datadir <-"../../data/"# "/scratch/PI/sph/resilience/"
curdir <- getwd()
source('../utils.R')

theme_set(theme_bw())
theme_update(
  text = element_text(15),
  panel.border = element_rect(size = 0.5),
  panel.grid = element_blank(),
  axis.ticks = element_blank(),
  legend.title = element_text(size = 12),
  legend.text = element_text(size = 12),
  legend.position = "bottom",
  strip.background = element_blank()
)
```


```{r, include = FALSE}
ps <- readRDS(file.path(datadir, "16S/phyloseq/perturb_physeq_fltr2438_15Jul19.rds"))
psSubj <- readRDS("../../data/16S/phyloseq/perturb_physeq_participants_decontam_15Jul19.rds")
```

Sample Information {#sample-info .unnumbered}
========================================


```{r, eval = FALSE}
# loading sample info
smpinfo_file <- file.path(
  datadir, "sample_info/Mapping_File_19Oct2018.xlsx") 

# Measurement Information
meas <- readxl::read_xlsx(smpinfo_file, sheet = "Meas")
colnames(meas) <- meas[1, ]
meas <- meas[-1, ] %>%
  rename(Samp_ID = SampID) %>%
  select(Meas_ID:Meas_Type)
```

```{r, eval = FALSE}
# Sample Information
interv_levs <- c("NoInterv", 
                 "PreDiet", "MidDiet", "PostDiet", 
                 "PreCC", "MidCC", "PostCC", 
                 "PreAbx", "MidAbx", "PostAbx", "UnpAbx")

samp <- readxl::read_xlsx(smpinfo_file, "Samp", skip = 1)  %>%
  mutate_at(
    .vars = c("Diet_Interval", "CC_Interval", "Abx_Interval"),
    .funs = function(x) {
      x <- ifelse(x == "Midabx","MidAbx", x)
      x <- ifelse(x == "NA", "NoInterv", x)
      x <- factor(x, interv_levs)
      return(x)
    }
  ) %>%
  mutate_at(
    .vars = vars(contains("Date")),
    .funs = function(x) {
      as.Date(as.numeric(x), origin="1899-12-30")
    }
  ) %>%
  mutate(
    Interval = paste0(Diet_Interval, "_", CC_Interval, "_", Abx_Interval),
    Interval = ifelse(grepl("MidAbx", Interval), "MidAbx", Interval),
    Interval = ifelse(grepl("MidDiet", Interval), "MidDiet", Interval),
    Interval = ifelse(grepl("MidCC", Interval), "MidCC", Interval),
    Interval = ifelse(!grepl("Post", Interval) & !grepl("Mid", Interval) &
                  !grepl("UnpAbx", Interval), "NoInterv", Interval),
    Interval = gsub("NoInterv_", "", Interval),
    Interval = gsub("_NoInterv", "", Interval),
    Interval = factor(Interval, unique(Interval))
  ) %>%
  mutate(
    Group = sapply(1:nrow(.), function(i) {
      group <- c()
      if(Diet_RelDay[i] != "NA") group <- c(group, "Diet")
      if(CC_RelDay[i] != "NA") group <- c(group, "CC")
      if(Abx_RelDay[i] != "NA") group <- c(group, "Abx")
      if(length(group) == 0) group <- "NoIntv"
      return(paste(group, collapse = "_"))
    })
  ) %>%
  select(Samp_ID:Abx_RelDay, Interval, Group) 
head(samp)
```

```{r, eval = FALSE}
# Subject Data
subj_data <- readxl::read_xlsx(
  smpinfo_file, sheet = "Subj", skip = 1, 
  col_types = c(rep("text", 5), rep("guess", 864))) %>%
  mutate_at(
    .vars = vars(contains("Date")),
    .funs = function(x) {
      as.Date(as.numeric(x), origin="1899-12-30")
    }
  ) %>%
  mutate(Age = `Age...9`) %>%
  dplyr::select(Subject, First_Sample_Date, CC_Date, Diet_StartDate, 
         Abx_StartDate, Age, contains("Age..."), Gender:Blood_Pressure, Blood_Type, BirthYear) 
subj_data <- subj_data[-1, ]
head(subj_data)
```

```{r eval = FALSE}
meas_info <- meas %>%
  left_join(samp, by = c("Samp_ID")) %>%
  left_join(subj_data, by = "Subject") %>%
  mutate(Event_Code = ifelse(!is.na(Event_Code.x), Event_Code.x, Event_Code.y))
write.csv(meas_info, file.path(datadir, "sample_info/meas_info_15July2019.csv"))
```

```{r, eval = FALSE}
meas_info <- read.csv( file.path(datadir, "sample_info/meas_info_15July2019.csv"), row.names = 1) %>%
  mutate_at(.vars = vars(contains("Date")), 
            .funs = function(x) {as.Date(as.character(x), "%Y-%m-%d")}) 
head(meas_info)
```

```{r, eval = FALSE}
start_dates <- meas_info %>%
   group_by(Subject, First_Sample_Date) %>%
    summarise(Real_First_Sample_Date = min(Samp_Date, na.rm = TRUE))
start_dates %>% filter(First_Sample_Date != Real_First_Sample_Date)
```

Fix one incorrect first sample date:

```{r eval = FALSE}
meas_info <- meas_info %>%
  left_join(start_dates) %>%
  mutate(DaysFromStart = as.numeric(Samp_Date - Real_First_Sample_Date ))
write.csv(meas_info,  file.path(datadir, "sample_info/meas_info_15July2019.csv"))
```


```{r}
meas_info <- read.csv( file.path(datadir, "sample_info/meas_info_15July2019.csv"), row.names = 1) 
smp_modality_count <- meas_info %>%
  select(Meas_ID, Samp_ID, Meas_Type) %>%
  group_by(Samp_ID) %>%
  summarise(
    modalities = paste0(unique(Meas_Type), collapse = "_"),
    modality_count = n()
  ) %>%
  mutate(
    modalities = ifelse(modalities == "MetaT_MetaG_16S", 
                        "16S_MetaG_MetaT", modalities)
  )
```

```{r exp-design, fig.width=10, fig.height=12}

df2plot <- meas_info %>%
  left_join(smp_modality_count) %>% 
  select(Meas_ID, Samp_ID, Subject, Samp_Date, Meas_Type, 
         Group, Interval, modalities, modality_count) %>%
  mutate(Samp_Date = as.Date(Samp_Date)) %>%
  filter(!grepl("NA", Subject))


new_intv_levs <- unique(meas_info$Interval)
colors <- colorRampPalette(brewer.pal(8, "Set3"))(length(new_intv_levs)-2)
colors <- c("grey57", "red", colors)
names(colors) <- c("NoInterv", "UnpAbx", 
                   setdiff(new_intv_levs, c("NoInterv", "UnpAbx")))
ggplot(df2plot %>% 
         filter(
           !(Group %in% c("Abx", "CC", "CC_Abx")),
           !is.na(Subject), !grepl("QC", Subject), 
           !(Subject  %in% c("DAG", "DAC")))
       ) +
  geom_point(
      pch = 124,
      aes(y = Subject, x = Samp_Date, color = Interval,
          fill = Interval, size = modality_count)
    ) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    scale_size_continuous(range = c(5, 10)) +
    theme(legend.direction = "horizontal", legend.box = "vertical",
          strip.text.y = element_text(angle=0)) +
    guides(color = guide_legend(ncols = 4, override.aes = list(size=5, pch = 16)),
           size = guide_legend(override.aes = list(color="black")))
```


```{r exp-design-all, fig.width=10, fig.height=15}
ggplot(df2plot) +
  geom_point(
      pch = 124,
      aes(y = Subject, x = Samp_Date, color = Interval,
          fill = Interval, size = modality_count)
    ) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    scale_size_continuous(range = c(5, 10)) +
    theme(legend.direction = "horizontal", legend.box = "vertical",
          strip.text.y = element_text(angle=0)) +
    guides(color = guide_legend(override.aes = list(size=5)),
           size = guide_legend(override.aes = list(color="black")))
```


```{r}
rm("colors", "df2plot", "smp_modality_count", "new_intv_levs")
```

```{r}
levels <- unique(meas_info$Interval)
cols_itv <- c("grey77", colorRampPalette(brewer.pal(9, "Set1"))(length(levels) - 1))
names(cols_itv) <- c("NoInterv", setdiff(levels, "NoInterv"))
```


Generate Phyloseq Object {#generate-phyloseq .unnumbered}
========================================

First we need to generate a new phyloseq object with samples
from two new batches.

We read the data stored as a `phyloseq` object from batches 1-4, 
then 5-6.

```{r get-physeq-batches-1-4, eval = FALSE}
# Bacthes 1-4
first_batches_dir <- file.path(datadir, "16S/16S_dada2_results/")

psRDP <- readRDS(file.path(first_batches_dir, "phyloseq_RDP.rds"))
psRDP

psRDPMult <- readRDS(file.path(first_batches_dir, "phyloseq_RDP_multispecies.rds"))
psRDPMult

psSilva <- readRDS(file.path(first_batches_dir, "phyloseq_Silva.rds"))
psSilva

psSilvaMult <- readRDS(file.path(first_batches_dir, "phyloseq_Silva_multispecies.rds"))
psSilvaMult
```

```{r get-physeq-batches-5-6, eval = FALSE}
# Bacthes 5-6
# Copied from curie server in folder:
# /relman02/data/relman/hmd/HMD_16S_V45/HMD_16S_5-6_annex/16S_5-6_DADA2_output/
second_batches_dir <- file.path(datadir, "16S/Les_Results/16S_5-6_DADA2_output")

psRDP2 <- readRDS(file.path(second_batches_dir, "phyloseq_RDP.rds"))
psRDP2

psRDPMult2 <- readRDS(file.path(second_batches_dir, "phyloseq_RDP_multispecies.rds"))
psRDPMult2

psSilva2 <- readRDS(file.path(second_batches_dir, "phyloseq_Silva.rds"))
psSilva2

psSilvaMult2 <- readRDS(file.path(second_batches_dir, "phyloseq_Silva_multispecies.rds"))
psSilvaMult2
```


```{r, eval = FALSE}
psRDP_merge <- merge_phyloseq(psRDP, psRDP2)
psRDPMult_merge <- merge_phyloseq(psRDPMult, psRDPMult2)
psSilva_merge <- merge_phyloseq(psSilva, psSilva2)
psSilvaMult_merge <- merge_phyloseq(psSilvaMult, psSilvaMult2)
rm(psRDP, psRDP2, psRDPMult, psRDPMult2, psSilva, psSilva2, psSilvaMult, psSilvaMult2)
```

```{r, eval = FALSE}
df <- bind_rows(
  data.frame(DB = "RDP", tax_table(psRDP_merge)@.Data),
  data.frame(DB = "Silva", tax_table(psSilva_merge)@.Data),
  data.frame(DB = "RDP_Mult", tax_table(psRDPMult_merge)@.Data),
  data.frame(DB = "Silva_Mult", tax_table(psSilvaMult_merge)@.Data)
)

# df %>%
#   group_by(DB) %>%
#   summarise_all(.funs = list(
#     function(x) sum(!is.na(x))
#   )
# )

# df %>%
#   group_by(DB) %>%
#   summarise_all(.funs = list(
#     function(x) round(100*sum(!is.na(x))/ntaxa(psRDP_merge), 1)
#   )
# )
```


```{r, eval = FALSE}
newTaxtab <- data.frame(as(tax_table(psRDP_merge), "matrix"))
newTaxtab <- cbind(newTaxtab, data.frame(SpeciesMult = as(tax_table(psRDPMult_merge), "matrix")[, "Species"]))
newTaxtab <- cbind(newTaxtab, data.frame(GenusSilva = as(tax_table(psSilva_merge), "matrix")[, "Genus"]))
newTaxtab <- cbind(newTaxtab, data.frame(SpeciesSilva = as(tax_table(psSilva_merge), "matrix")[, "Species"]))
newTaxtab <- cbind(
  newTaxtab, "SpeciesSilvaMult" = as(tax_table(psSilvaMult_merge), "matrix")[, "Species"])
newTaxtab <- cbind(newTaxtab, Seq = rownames(newTaxtab))
rownames(newTaxtab) <- paste0("Seq", 1:nrow(newTaxtab))

seqtab <- as(otu_table(psRDP_merge), "matrix")
seqtab <- seqtab[newTaxtab[, "Seq"], ]
rownames(seqtab) <- rownames(newTaxtab) 
```

```{r, eval=FALSE}
ps_smp_names <- colnames(seqtab) 
names(ps_smp_names) <- colnames(seqtab) 

batch_info <- gsub("HMD_", "", ps_smp_names)
batch_info <- gsub("\\_(.*)", "", batch_info)

ps_smp_names <- gsub("\\-(.*)", "", ps_smp_names)
ps_smp_names <- gsub("(.*)\\_M", "M", ps_smp_names)
head(ps_smp_names)
tail(ps_smp_names)
length(ps_smp_names) == length(unique(ps_smp_names))
```

As we see above, the id's are unique in our sample. Now, we check that
all the identifiers are included in the mapping file:

```{r, eval=FALSE}
all(ps_smp_names %in% meas_info$Meas_ID)
```

```{r, eval=FALSE}
ps_smp_names[which(!ps_smp_names %in% meas_info$Meas_ID)]
# "HMD_16S5_SplitLib_Unassigned" 
```


```{r, eval=FALSE}
keep_smp_names <- ps_smp_names[which(ps_smp_names %in% meas_info$Meas_ID)]
batch_info <- batch_info[which(ps_smp_names %in% meas_info$Meas_ID)]
names(batch_info) <- keep_smp_names
seqtab <- seqtab[, names(keep_smp_names)]
colnames(seqtab) <- keep_smp_names
```


We extract the information for the obtained samples:

```{r, eval=FALSE}
samp_16S <- meas_info %>%
  filter(Meas_ID %in% keep_smp_names) %>%
  left_join(data.frame(Meas_ID = names(batch_info), 
                       Sequencing_Run = batch_info))
rownames(samp_16S) <- samp_16S$Meas_ID
head(samp_16S)
```

```{r, eval=FALSE}
ps_combined <- phyloseq(
  otu_table(seqtab, taxa_are_rows = TRUE), 
  sample_data(samp_16S),
  tax_table(as(newTaxtab, "matrix")))
ps_combined
```

```{r, eval=FALSE}
physeqfile <- file.path(datadir, "16S/phyloseq/perturb_physeq_15Jul19.rds")
saveRDS(ps_combined, file = physeqfile)
# otu_table()   OTU Table:         [ 19638 taxa and 4561 samples ]
# sample_data() Sample Data:       [ 4561 samples by 39 sample variables ]
```



Preprocess Data {#preprocess .unnumbered}
=========================================


```{r, eval = FALSE}
physeqfile <- file.path(datadir, "16S/phyloseq/perturb_physeq_15Jul19.rds")
ps0 <- readRDS(file = physeqfile)
SMP <- data.frame(ps0@sam_data, stringsAsFactors = FALSE) %>%
  mutate(SampleDepth = sample_sums(ps0))
ps0

# otu_table()   OTU Table:         [ 19638 taxa and 4561 samples ]
# sample_data() Sample Data:       [ 4561 samples by 39 sample variables ]
# tax_table()   Taxonomy Table:    [ 19638 taxa by 12 taxonomic ranks ]
```



We will now remove samples from subject who did not have sufficient number of
samples sequenced, either because they came from a different experiment or
because the subjects did not complete the study.

```{r, eval = FALSE}
freq <- table(ps0@sam_data$Subject)
remove_subjects <- names(freq)[freq < 10] # "AAA" "AAB" "AAN" "DAC" "DAG"
remove_subjects
ps <- subset_samples(ps0, !Subject %in% remove_subjects)
ps
# otu_table()   OTU Table:         [ 19638 taxa and 4521 samples ]
```


```{r distr-seq-samples, fig.width=10, fig.height=10}

df2plot <- data.frame(sample_data(psSubj)) %>%
  select(Meas_ID, Samp_ID, Subject, Samp_Date, Meas_Type, 
         Group, Interval) %>%
  mutate(Samp_Date = as.Date(Samp_Date))


new_intv_levs <- unique(df2plot$Interval)
colors <- colorRampPalette(brewer.pal(8, "Set3"))(length(new_intv_levs)-2)
colors <- c("grey57", "red", colors)
names(colors) <- c("NoInterv", "UnpAbx", 
                   setdiff(new_intv_levs, c("NoInterv", "UnpAbx")))
df2plot %>% 
  ggplot() +
  geom_point(
      pch = 124, size = 3,
      aes(y = Subject, x = Samp_Date, color = Interval,
          fill = Interval)
    ) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    theme(legend.direction = "horizontal", legend.box = "vertical",
          strip.text.y = element_text(angle=0)) +
    guides(color = guide_legend(ncols = 4, override.aes = list(size=5, pch = 16)),
           size = guide_legend(override.aes = list(color="black")))
```

Filter taxa {#filter .unnumbered} 
---------------------------------

Next, we will filter out taxa which are not present in samples of at least 
two different subjects.

```{r, eval = FALSE}
minTaxaSubjPrev <- 2
minNonZeroSamples <- 5
minPrevLevel <- 5 

# Compute prevalence across subjects
num_subj_present <- function(physeq, subject_indicator = "Subject") {
  smp <- data.frame(physeq@sam_data, stringsAsFactors = FALSE)
  seqtab <- as(otu_table(physeq), "matrix")
  if(!taxa_are_rows(physeq)) seqtab <- t(seqtab)
  sapply(1:ntaxa(physeq), function(i) {
    subj.present <- smp[[subject_indicator]][seqtab[i, ] >= minPrevLevel]
    freq <- table(subj.present)
    sum(freq >= minNonZeroSamples)
    #length(unique(subj.present))aDAMAD
  })
}
subjects_prevalence <- num_subj_present(ps)

# Filter taxa present in more than 1 subject
ps <- subset_taxa(ps, subjects_prevalence >= minTaxaSubjPrev)
ps
#otu_table()   OTU Table:         [ 2438 taxa and 4521 samples ]
```

```{r,  eval = FALSE}
ps <- subset_samples(ps, sample_sums(ps) > 0)
ps
#otu_table()   OTU Table:         [ 2438 taxa and 4500 samples ]
saveRDS(ps, file = file.path(datadir, "16S/phyloseq/perturb_physeq_fltr2438_15Jul19.rds"))
```



Estimate the phylo tree {#phylo-tree .unnumbered} 
-------------------------------------------------

Go to code in 
"/scratch/users/lanhuong/Projects/perturbation_16s/16S/construct_phylo_tree.R"
and see how the tree was estimated.



```{r, eval = FALSE}
ps <- readRDS(file.path(datadir, "16S/phyloseq/perturb_physeq_fltr2438_15Jul19.rds"))
ps
# otu_table()   OTU Table:         [ 2438 taxa and 4500 samples ]
load(file.path(datadir, "16S/phyloseq/phylo_fit_fltr2438_15Jul19.rda"))
phy_tree(ps) <- rootedGTRtree
ps
saveRDS(ps, file = file.path(datadir, "16S/phyloseq/perturb_physeq_fltr2438_15Jul19.rds"))
```


```{r}
# Assign colors to taxa
tax_rank <- "Family"
n_taxa <- 15

taxTab <- data.frame(as(tax_table(ps), "matrix"))

taxFreq <- data.frame(Seq_ID = taxa_names(ps), Seq_Sum = taxa_sums(ps)) %>%
  left_join(taxTab %>%
              rownames_to_column("Seq_ID")) %>%
  group_by_(tax_rank) %>%
  summarise(Taxa_Sum = sum(Seq_Sum)) %>%
  filter(!is.na(.[[tax_rank]])) %>%
  arrange(desc(Taxa_Sum)) %>%
  top_n(n_taxa) %>%
  mutate(color = colorRampPalette(brewer.pal(8, "Dark2"))(n_taxa))

other <-  data.frame("Other", 0, "grey77")
colnames(other) <- colnames(taxFreq)
taxFreq <- rbind(taxFreq, other)

taxa_col <- tibble(
  Seq_ID = taxa_names(ps),
  Taxa = c(tax_table(ps)[, tax_rank])) %>%
  mutate(
    Taxa = ifelse(is.na(Taxa), "Unknown", Taxa),
    Taxa = ifelse(Taxa == "", "Unknown", Taxa),
    Taxa = ifelse(Taxa %in% taxFreq[[tax_rank]], Taxa, "Other")) %>%
  left_join(taxFreq, by=c("Taxa" = tax_rank))
```

```{r phylotree, fig.height=10, fig.width=15}
default_par <- par()
par(mar=c(7,1,1,1), xpd = TRUE)
plot(ps@phy_tree, label.offset=0.7, show.tip.label = FALSE, 
     direction = "downwards") 
par(mar=c(1,1,1,1), xpd = TRUE)
ape::tiplabels(pch=21, col="black", adj=1, bg=taxa_col$color, cex=1) 
legend(x= 600, y = -0.1, legend = taxFreq[[tax_rank]], 
       fill = taxFreq$color, ncol = 4, cex = 1.2)
```




Decontamination {#decontam .unnumbered} 
-------------------------------------------------


Remove contamination:


```{r}
# devtools::install_github("benjjneb/decontam")
library(decontam); packageVersion("decontam")
df <- data.frame(sample_data(ps)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=Subject == "NA_QC")) + 
  geom_point() +
  scale_y_log10()
```

```{r}
table(df$Subject)
```


```{r}
df %>% filter(Subject == "NA_QC") %>% 
  .[["Event_Code"]] %>% as.character 
```


```{r}
df %>% filter(Subject == "NA_QC") %>% .[["Samp_Type"]] %>% table()
```

```{r}
sample_data(ps)$neg_control <- (sample_data(ps)$Samp_Type == "ExtrControl" | sample_data(ps)$Samp_Type == "ExtrCont")
df$neg_control <- df$Samp_Type == "ExtrControl" | df$Samp_Type == "ExtrCont"
```

```{r}
df %>% filter(!neg_control) %>%
  .[["LibrarySize"]] %>% summary()
```

```{r}
df %>% filter(neg_control) %>%
  .[["LibrarySize"]] %>% summary()
```


```{r}
contamdf.prev <- isContaminant(ps, method="prevalence", neg="neg_control")
table(contamdf.prev$contaminant)
```

```{r}
# contaminants
tax_table(ps)[which(contamdf.prev$contaminant), 2:7]
```


```{r}
contamdf.prev05 <- isContaminant(
  ps, method="prevalence", neg="neg_control", threshold=0.5)
table(contamdf.prev05$contaminant)
```

```{r}
# contaminants
tax_table(ps)[which(contamdf.prev05$contaminant)[1:10], 2:7]
```

```{r, eval  = FALSE}
psFamily <- tax_glom(ps, taxrank = "Family")
family_sum <- taxa_sums(psFamily)
topFamily <- as(psFamily@tax_table, "matrix")[order(-family_sum)[1:10], "Family"]
saveRDS(psFamily, file = "output/perturb_physeq_glom_family_15Jul19.rds")
```

```{r, inlcude  = FALSE}
psFamily <- readRDS(file = "output/perturb_physeq_glom_family_15Jul19.rds")
family_sum <- taxa_sums(psFamily)
topFamily <- as(psFamily@tax_table, "matrix")[order(-family_sum)[1:10], "Family"]
```


```{r}
ps.pa <- transform_sample_counts(ps, function(abund) 1*(abund>0))
ps.pa.neg <- subset_samples(ps.pa, sample_data(ps.pa)$neg_control)
ps.pa.pos <- subset_samples(ps.pa, !sample_data(ps.pa)$neg_control)
ps.pa.pos.mid <- subset_samples(
  ps.pa.pos, (Diet_Interval == "MidDiet" | Abx_Interval == "MidAbx" | CC_RelDay %in% seq(1, 5)))

# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(
  taxa.sum = taxa_sums(ps),
  pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
  contaminant=contamdf.prev$contaminant)

df.pa.mid <- data.frame(
  taxa.sum = taxa_sums(ps),
  pa.pos=taxa_sums(ps.pa.pos.mid), 
  pa.neg=taxa_sums(ps.pa.neg),
  contaminant=contamdf.prev$contaminant)
```

```{r, fig.width=8, fig.height=5}
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (Subject Samples)") +
  theme(legend.position = "right") 
```

```{r, fig.width=8, fig.height=6}
df.pa <- df.pa %>%
  rownames_to_column("SeqID") %>%
  mutate(
    Family = factor(as(tax_table(ps.pa.pos), "matrix")[, "Family"], levels = topFamily)
  ) 

df.pa %>%  
ggplot(aes(x=pa.neg, y=pa.pos, color=Family)) +
  geom_point(data = df.pa %>%  filter(pa.neg <= 25)) + 
  geom_text(data = df.pa %>%  filter(pa.neg > 25), aes(label = SeqID)) +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (Subject Samples)") +
  theme(legend.position = "right") + theme(legend.position = "bottom") +
  guides(color=guide_legend(ncol=3))

```

```{r, fig.width=8, fig.height=5}
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=asinh(taxa.sum))) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (Subject Samples)") +
  scale_color_viridis_c() + theme(legend.position = "right") 
```

```{r}
cross_contam_seqs <- df.pa %>% filter(pa.neg > 25) %>% .[["SeqID"]]
knitr::kable(tax_table(ps)[cross_contam_seqs, 5:7])
```


```{r, fig.width=8, fig.height=6}
ggplot(data=df.pa.mid, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (Mid Perturbation Samples)") 
```

```{r, fig.width=8, fig.height=6}
df.pa.mid <- df.pa.mid %>%
  rownames_to_column("SeqID") %>%
  mutate(
    Family = factor(as(tax_table(ps.pa.pos.mid), "matrix")[, "Family"], levels = topFamily)
  ) 

df.pa.mid %>%  
ggplot(aes(x=pa.neg, y=pa.pos, color=Family)) +
  geom_point(data = df.pa.mid %>%  filter(pa.neg <= 25)) + 
  geom_text(data = df.pa.mid %>%  filter(pa.neg > 25), aes(label = SeqID)) +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (Subject Samples)") +
  theme(legend.position = "right") + theme(legend.position = "bottom") +
  guides(color=guide_legend(ncol=3))


```

```{r, fig.width=8, fig.height=5}
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=asinh(taxa.sum))) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (Mid Perturbatin Samples)") +
  scale_color_viridis_c() + theme(legend.position = "right") 
```

```{r}
cross_contam_seqs_mid_perturb <- df.pa.mid %>% 
  filter(pa.neg > 25, pa.pos > 300) %>% .[["SeqID"]]
knitr::kable(tax_table(ps)[cross_contam_seqs_mid_perturb, 5:7])
```

```{r}
contaminant_seqs <- rownames(contamdf.prev[contamdf.prev$contaminant, ])
ps <- subset_taxa(ps, !taxa_names(ps) %in% contaminant_seqs)
saveRDS(ps, file = file.path(datadir, "16S/phyloseq/perturb_physeq_decontam_15Jul19.rds"))

save(list=c("cross_contam_seqs_mid_perturb", "cross_contam_seqs", "contamdf.prev", "contamdf.prev05"),
     file = "output/potential_cross_contam.rda")
```



Filter samples {#filter .unnumbered} 
---------------------------------


We then separate the subject data and the negative controls:

```{r, eval = FALSE}
psControls <- subset_samples(ps, Subject == "NA_QC")
psControls
# otu_table()   OTU Table:         [ 2425 taxa and 84 samples ]
```

```{r, eval = FALSE}
psSubj <- subset_samples(ps, Subject != "NA_QC")
psSubj
# otu_table()   OTU Table:         [ 2425 taxa and 4416 samples ]

minSampleSum <- 1e4
psSubj <- subset_samples(psSubj, sample_sums(psSubj) > minSampleSum)
psSubj
# otu_table()   OTU Table:         [ 2425 taxa and 4402 samples ]
```


```{r eval = FALSE}
saveRDS(psControls, file = file.path(datadir, "16S/phyloseq/perturb_physeq_controls_decontam_15Jul19.rds"))
saveRDS(psSubj, file = file.path(datadir, "16S/phyloseq/perturb_physeq_participants_decontam_15Jul19.rds"))
```



Variance Stabilization {#vst .unnumbered} 
-------------------------------------------------

```{r, eval = FALSE}
# Transform counts to not "over-weight" highly abundant taxa
#ps_asinh <- transform_sample_counts(ps, function(x) {asinh(x)})
```

```{r, eval = FALSE}
# Takes too long
dds <- phyloseq_to_deseq2(ps, design = ~ 1)
dds <- estimateSizeFactors(dds, type = "poscounts")
norm_counts <- counts(dds, normalized = TRUE)

# ## The following takes too long; and vst() does not work with too few genes with rowmeans > 5
# vsd <- varianceStabilizingTransformation(dds)
# ps_vst <- ps
# otu_table(ps_vst) <- otu_table(assay(vsd), taxa_are_rows = TRUE)
# save(list = c("ps", "pscontrols", "ps_asinh", "SMP", "SUBJ", "dds",
#               "norm_counts", "ps_vst", "vsd"), 
#      file = processed_datafile)
```



Distance Matrices {#dist .unnumbered} 
========================================


```{r}
SMP <- data.frame(psSubj@sam_data, stringsAsFactors = FALSE) %>%
  mutate(SampleDepth = sample_sums(psSubj))
  
SUBJ <-  SMP %>%
  select(Subject, Group, Age:BMI) %>%
  distinct()
```


```{r, eval = FALSE}
# takes a while
brayD.ihs <- phyloseq::distance(
    transform_sample_counts(psSubj, function(x) {asinh(x)}), 
    method = "bray")
save(list = c("brayD.ihs"), file = "output/pairwise_dist_subj_16S.rda")
```

```{r, eval = FALSE}
jaccardD <- phyloseq::distance(
    psSubj, method = "jaccard", binary = TRUE)
save(list = c("brayD.ihs", "jaccardD"), file = "output/pairwise_dist_subj_16S.rda")
```

```{r, eval = FALSE}
study_arms <- as.character(unique(psSubj@sam_data$Group))

uniFracD.lst <- list()
study_arms <- setdiff(study_arms, names(uniFracD.lst))
for(gr in study_arms) {
  ps.gr <- phyloseq::subset_samples(psSubj, Group == gr)
  print(ps.gr)
  uniFracD.lst[[gr]] <- phyloseq::distance(
    ps.gr, method = "unifrac", parallel = TRUE, 
    fast = TRUE, normalized = TRUE, weighted = FALSE)
  save(list = c("brayD.ihs", "jaccardD", "uniFracD.lst"), 
       file = "../analysis/analysis_summer2019/output/pairwise_dist_subj_16S.rda")
}

save(list = c("brayD.ihs", "jaccardD", "uniFracD.lst"), 
     file = "../analysis/analysis_summer2019/output/pairwise_dist_subj_16S.rda")
```



Process distance matrices

```{r}
distMat_to_2long <- function(distMat, sample_info = NULL, 
                             cols_to_keep = NULL,
                             sample_id = "Meas_ID",
                             keep_only_same_subj = TRUE) {
  dist.df <- reshape2::melt(
    as.matrix(distMat), 
    varnames = c("Sample_1", "Sample_2"),
    value.name = "dist") %>%
    mutate(Sample_1 = as.character(Sample_1),
           Sample_2 = as.character(Sample_2))
  if(!is.null(sample_info) & sample_id %in% colnames(sample_info)) {
    vars <- c(sample_id)
    names(vars) <- "Meas_ID"
    if(!is.null(cols_to_keep)) {
      vars <- c(vars, cols_to_keep)
      names(vars) <- c("Meas_ID", cols_to_keep)
    }
    dist.df <- dist.df %>%
      left_join(
        sample_info %>% select(!!vars), 
        by = c("Sample_1" = "Meas_ID"))
    
    dist.df <- dist.df %>%
      left_join(
        sample_info %>% select(!!vars), 
        by = c("Sample_2" = "Meas_ID"), 
        suffix = c("_1", "_2"))
    
    if(keep_only_same_subj) {
      dist.df <- dist.df %>%
        filter(Subject_1 == Subject_2) %>%
        rename(Subject = Subject_1) %>%
        select(-Subject_2)
    }
  }
  return(dist.df)
}

```


```{r, eval = FALSE}
bray.df <- distMat_to_2long(
  brayD.ihs, SMP, 
  cols_to_keep = c("Subject", "Group", "DaysFromStart", 
                   SMP %>% select(contains("RelDay")) %>% colnames()))  %>%
  rename(Group = Group_1) %>%
  select(-Group_2)
dim(bray.df)

jacc.df <- distMat_to_2long(
  jaccardD, SMP, 
  cols_to_keep = c("Subject", "Group", "DaysFromStart", 
                   SMP %>% select(contains("RelDay")) %>% colnames())) %>%
  rename(Group = Group_1) %>%
  select(-Group_2)
dim(jacc.df)


uniFrac.df <- list()
for(gr in names(uniFracD.lst)){
  uniFrac.df[[gr]] <- distMat_to_2long(
    uniFracD.lst[[gr]], SMP, 
    cols_to_keep = c("Subject", "Group", "DaysFromStart", 
                     SMP %>% select(contains("RelDay")) %>% colnames())) %>%
    rename(Group = Group_1) %>%
    select(-Group_2)
}
uniFrac.df <- do.call("rbind", uniFrac.df)
dim(uniFrac.df)


# [1] 304740     13
# [1] 304740     13
save(list = c("brayD.ihs", "bray.df", "jaccardD", "jacc.df",
              "uniFracD.lst", "uniFrac.df"),
     file = "output/pairwise_dist_subj_16S.rda")

```



```{r}
sessionInfo()
```

