---
title: "Presentation Plots"
author: 
  - name: Lan Huong Nguyen
    affiliation: Institute for Computational and Mathematical Engineering, Stanford University, CA 94305
keywords: microbiome, metatranscriptomics.
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{Gene expression time course data analysis}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis"}
rm(list = ls())
library("BiocStyle")
library("rmarkdown")
options(width = 200, stringsAsFactors = FALSE) 
knitr::opts_chunk$set(
  message = FALSE, error = FALSE, warning = FALSE, 
  fig.width = 8, fig.height = 6,
  fig.path = "./figs/relman_group_presentation/", 
  dev='png') 
```

```{r setup}
library("tidyverse")
library("readxl")
library("data.table")
library("gridExtra")
library("RColorBrewer")
library("feather")

theme_set(theme_bw())
theme_update(
  text = element_text(15),
  panel.border = element_rect(size = 0.5),
  panel.grid = element_blank(),
  axis.ticks = element_blank(),
  legend.title = element_text(size = 10),
  legend.text = element_text(size = 10),
  legend.position = "bottom",
  strip.background = element_blank()
)
```

Study Design {#study-design .unnumbered}
========================================


```{r}
# loading sample info
datadir <- "/scratch/PI/sph/resilience/"
smpinfo_file <- file.path(datadir, "sample_info/Mapping_Files_26March2018.xlsx") 

# Measurement Information
meas <- read_xlsx(smpinfo_file, sheet = "Meas")
colnames(meas) <- meas[1, ]
meas <- meas[-1, ] %>%
  rename(Samp_ID = SampID) %>%
  select(Meas_ID:Meas_Type)

```

```{r}
# Sample Information
interv_levs <- c("NoInterv", 
                 "PreDiet", "MidDiet", "PostDiet", 
                 "PreCC", "MidCC", "PostCC", 
                 "PreAbx", "MidAbx", "PostAbx", "UnpAbx")

samp <- read_xlsx(smpinfo_file, "Samp", skip = 1) 
samp <- samp %>%
  filter(Samp_Type != "ExtrCont") %>%
  mutate(
    Diet_Interval = ifelse(Diet_Interval == "NA", "NoInterv", Diet_Interval),
    CC_Interval = ifelse(CC_Interval == "NA", "NoInterv", CC_Interval),
    Abx_Interval = ifelse(Abx_Interval == "NA", "NoInterv", Abx_Interval),
    Abx_Interval = ifelse(Abx_Interval == "Midabx","MidAbx", Abx_Interval),
    Diet_Interval = factor(Diet_Interval, interv_levs),
    CC_Interval = factor(CC_Interval, interv_levs),
    Abx_Interval = factor(Abx_Interval, interv_levs),
    Interval = paste0(Diet_Interval, "_", CC_Interval, "_", Abx_Interval)) %>%
  select(Samp_ID:Abx_Interval, Interval) 
  

samp <- samp %>%
  mutate(
    Interval = gsub("NoInterv_NoInterv_NoInterv", "NoInterv", Interval),
    Interval = gsub("NoInterv_", "", Interval),
    Interval = gsub("_NoInterv", "", Interval),
    Interval = ifelse(!grepl("Post", Interval) & !grepl("Mid", Interval) &
                      !grepl("UnpAbx", Interval),
                      "NoInterv", Interval),
    Interval = ifelse(grepl("MidAbx", Interval), "MidAbx", Interval),
    Interval = ifelse(grepl("MidDiet", Interval), "MidDiet", Interval),
    Interval = ifelse(grepl("MidCC", Interval), "MidCC", Interval)
  ) 

new_intv_levs <- unique(samp$Interval) 
samp <- samp %>%
  mutate(
    Interval = factor(Interval, new_intv_levs)
  )

# Fix date issues
samp <- samp %>% 
  mutate_at(.funs = function(x) {as.Date(as.numeric(x), origin="1899-12-30")},
              .vars = select(samp, contains("Date")) %>% colnames())


```

```{r}
# Subject Data
subj_data <- read_excel(smpinfo_file,
                        sheet = "SubjNoRagged", skip = 1, 
                        col_types = c("text", rep("date", 4), rep("guess", 864)))  %>%
  mutate(CC = ifelse(!is.na(CC_Date), "CC", NA),
         Diet = ifelse(!is.na(Diet_StartDate), "Diet", NA),
         Abx = ifelse(!is.na(Abx_StartDate), "Abx", NA),
         Group = sapply(1:nrow(.), function(i) {
           z <- c(CC[i], Diet[i], Abx[i])
           z <- paste(z[!is.na(z)], collapse = "_")
           if(z == "") z <- "NoIntv"
           return(z)
           })) %>%
  select(Subject, Group, First_Sample_Date, CC_Date, Diet_StartDate, Abx_StartDate,
         Age:BirthYear)
head(subj_data)
```

```{r}
meas_info <- meas %>%
  left_join(samp) %>%
  left_join(subj_data) %>%
  filter(
    !startsWith(Subject, "AA"),
    Meas_Type %in% c("16S", "MetaG", "MetaT"))
```


```{r}
smp_modality_count <- meas_info %>%
  select(Meas_ID, Samp_ID, Meas_Type) %>%
  group_by(Samp_ID) %>%
  summarise(
    modalities = paste0(unique(Meas_Type), collapse = "_"),
    modality_count = n()
  ) %>%
  mutate(
    modalities = ifelse(modalities == "MetaT_MetaG_16S", "16S_MetaG_MetaT", modalities)
  )
  
df2plot <- meas_info %>%
  left_join(smp_modality_count) %>% 
  select(Meas_ID, Samp_ID, Subject, Samp_Date, Meas_Type, 
         Group, Interval, modalities, modality_count) %>%
  filter(!grepl("NA", Subject))
```

```{r exp-design, fig.width=10, fig.height=15}
colors <- colorRampPalette(brewer.pal(8, "Set3"))(length(new_intv_levs)-2)
colors <- c("grey57", "red", colors)
names(colors) <- c("NoInterv", "UnpAbx", 
                   setdiff(new_intv_levs, c("NoInterv", "UnpAbx")))

ggplot(df2plot) +
  geom_point(
      pch = 124,
      aes(y = Subject, x = Samp_Date, color = Interval,
          fill = Interval, size = modality_count)
    ) +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    scale_color_manual(values = colors) +
    scale_size_continuous(range = c(5, 10)) +
    theme(legend.direction = "horizontal", legend.box = "vertical") +
    guides(color = guide_legend(override.aes = list(size=5)),
           size = guide_legend(override.aes = list(color="black")))
```

```{r exp-design-group, fig.width=10}
for(gr in unique(df2plot$Group)) {
  plt <- ggplot(df2plot %>% filter(Group == gr)) +
    geom_point(
      pch = 124,
      aes(y = Subject, x = Samp_Date, color = Interval,
          fill = Interval, size = modality_count)
    ) +
    #facet_grid(Group ~ ., scales = "free", space = "free") +
    #scale_fill_manual(values = colors) +
    scale_color_manual(values = colors) +
    scale_size_continuous(range = c(5, 10)) +
    theme(legend.direction = "horizontal", legend.box = "vertical") +
    guides(color = guide_legend(override.aes = list(size=5)),
           size = guide_legend(override.aes = list(color="black")))
  print(plt + ggtitle(gr))
}

```


Read Coverage {#read-coverage .unnumbered}
========================================

Here we summarize the read coverage statistics for all data modalities.

16S {#read-coverage-16S .unnumbered}
------------------------------------

```{r}
stats16sfile <- file.path(datadir, "16S/16S_dada2_results/track_all_pools.csv")
stats16s <- read.csv(stats16sfile, row.names = 1, stringsAsFactors = FALSE) %>%
  mutate(Meas_ID = gsub(".*_M", "M", sample.id),
         Meas_ID = gsub("\\-.*", "", Meas_ID)) %>%
  select(pool, Meas_ID, sample.id:nonchim) %>%
  rename(batch = pool)
head(stats16s)
```

```{r}
physeqfile <- file.path(datadir, "16S/phyloseq/perturb_physeq_22Jan18.rds")
ps <- readRDS(physeqfile)
ps
```


```{r}
smp_depth_16s <- data.frame(
  physeq = sample_sums(ps),
  annotated_genus = sample_sums(subset_taxa(ps, !is.na(Genus))),
  annotated_species = sample_sums(subset_taxa(ps, !is.na(Species)))) %>%
  rownames_to_column("Meas_ID")

stats16s <- stats16s %>%
  left_join(smp_depth_16s)
```


```{r}
all_reads <- colSums(stats16s %>% select(-(batch:file.name)))
(total_reads_coverage_16s <- 100*all_reads/all_reads[1])
```


```{r}
percent_stats16s <- stats16s %>% 
  mutate_at(
    .funs = function(x) 100 * x/(.)$input,
    .vars = setdiff(colnames((.)), c("batch", "Meas_ID","sample.id", "file.name"))
  ) 

reads_coverage_16s <- percent_stats16s %>%
  select(-(Meas_ID:file.name)) %>%
  group_by(batch) %>%
  summarize_all(median) %>%
  as.data.frame() %>%
  bind_rows(
    data.frame(
      batch = "All_16S", 
      percent_stats16s %>%
          select(-(batch:file.name)) %>%
          summarise_all(median),
      stringsAsFactors = FALSE
    )
  )
reads_coverage_16s %>%
  select(-tabled.rightlen, -nonchim) %>% # the same as merged and physeq
  mutate_at(.funs = function(x) round(x, digits = 1),
            .vars = vars(-batch))
```


MetaGenomics {#read-coverage-metag .unnumbered}
------------------------------------

```{r}
metagReadfile <- file.path(datadir, "metagenomics/merged/gene_reads.feather")
metagreads <- read_feather(metagReadfile) %>%
  mutate(genus = str_extract(species, "[^_]+")) %>%
  select( genus, species, gene_id, starts_with("M"))
metagreads[is.na(metagreads)] <- 0 
```


```{r}
statsMetaGfile <- file.path(datadir, "metagenomics/stats/all_samples_stats.csv")
statsMetaG <- read.csv(statsMetaGfile, stringsAsFactors = FALSE) %>%
  mutate(input = input_fwd) %>%
  select(Meas_ID, batch, input, input_fwd:annotated) %>%
  filter(startsWith(Meas_ID, prefix = "M")) %>%
  mutate(
    batch = gsub(".*\\/", "", batch),
    batch = gsub("\\.(.*)", "", batch)
)
```

```{r}
smp_depth_metag <- data.frame(
  metag_pangenome = colSums(metagreads %>% select(-genus, -species, -gene_id))
  ) %>%
  rownames_to_column("Meas_ID")

statsMetaG <- statsMetaG %>%
  left_join(smp_depth_metag) %>%
  rename(metag_marker_genes = annotated)
```

```{r}
statsMetaG %>% filter(is.na(metag_pangenome))
statsMetaG <- statsMetaG %>%
  mutate(input = input_fwd) %>%
  filter(!is.na(metag_pangenome))
```


```{r}
all_reads <- colSums(statsMetaG %>% select(-Meas_ID, -batch))
(total_reads_coverage_metaG <- 100*all_reads/all_reads[1])
```


```{r}
percent_statsMetaG <- statsMetaG %>% 
  mutate_at(
    .funs = function(x) 100 * x/(.)$input_fwd,
    .vars = setdiff(colnames((.)), c("batch", "Meas_ID"))
  ) 

reads_coverage_metaG <- percent_statsMetaG %>%
  select(-(Meas_ID)) %>%
  group_by(batch) %>%
  summarize_all(median) %>%
  as.data.frame() %>%
  bind_rows(
    data.frame(
      batch = "All_MetaG", 
      percent_statsMetaG %>%
          select(-batch, -Meas_ID) %>%
          summarise_all(median),
      stringsAsFactors = FALSE
    )
  ) %>%
  select(batch, input, starts_with("metag_"))

reads_coverage_metaG %>%
  mutate_at(.funs = function(x) round(x, digits = 1),
            .vars = vars(-batch))
```


MetaTranscriptomics {#read-coverage-metat .unnumbered}
------------------------------------

```{r}
statsMetaTfile <- file.path(datadir, "metatranscriptomics/stats/all_samples_stats.csv")
statsMetaT <- read.csv(statsMetaTfile, stringsAsFactors = FALSE) %>%
  mutate( input = input_fwd) %>%
  select(Meas_ID, batch, input, input_fwd:qual_fltr, mRNA, dmnd_refseq, dmnd_seed)
head(statsMetaT)
```

```{r}
all_reads <- colSums(statsMetaT %>% select(-Meas_ID, -batch))
(total_reads_coverage_metaT <- 100*all_reads/all_reads[1])
```


```{r}
percent_statsMetaT <- statsMetaT %>% 
  mutate_at(
    .funs = function(x) 100 * x/(.)$input_fwd,
    .vars = setdiff(colnames((.)), c("batch", "Meas_ID"))
  ) 

reads_coverage_metaT <- percent_statsMetaT %>%
  select(-(Meas_ID)) %>%
  group_by(batch) %>%
  summarize_all(median) %>%
  as.data.frame() %>%
  bind_rows(
    data.frame(
      batch = "All_MetaT", 
      percent_statsMetaT %>%
          select(-batch, -Meas_ID) %>%
          summarise_all(median),
      stringsAsFactors = FALSE
    )
  ) %>%
  select(batch, input, trimmed:dmnd_seed)

reads_coverage_metaT %>%
  mutate_at(.funs = function(x) round(x, digits = 1),
            .vars = vars(-batch))
```

Combine Data {#read-coverage-all .unnumbered}
------------------------------------

```{r}
cols <- c("batch", "Meas_ID", "input")
stats_input <- bind_rows(
  data.frame(modality = "16S", stats16s[, cols]),
  data.frame(modality = "MetaG", statsMetaG[, cols]),
  data.frame(modality = "MetaT", statsMetaT[, cols])
)
```

```{r, input-read-counts}
ggplot(stats_input, 
       aes(x = modality, y = input)) +
  geom_boxplot() +
  geom_jitter(width = 0.2, alpha = 0.3) 
```


```{r}
cats <- c("16S_physeq", "16S_annotated_genus", "16S_annotated_species",
          "MetaG_metag_pangenome","MetaG_metag_marker_genes", #"MetaT_mRNA", 
          "MetaT_dmnd_refseq", "MetaT_dmnd_seed")         
reads_coverage <- bind_rows(
  reads_coverage_16s %>%
    select(batch, physeq, annotated_genus, annotated_species) %>%
    gather(key = type, value = percentage_of_input, -batch) %>%
    mutate(modality = "16S"),
  reads_coverage_metaG %>%
    select(batch, metag_marker_genes, metag_pangenome) %>%
    gather(key = type, value = percentage_of_input, -batch) %>%
    mutate(modality = "MetaG"),
  reads_coverage_metaT %>%
    select(batch,  dmnd_refseq, dmnd_seed) %>% #mRNA,
    gather(key = type, value = percentage_of_input, -batch) %>%
    mutate( modality = "MetaT")) %>%
  filter(grepl("All", batch)) %>%
  mutate(
    category = paste0(modality, "_", type),
    category = factor(category, levels = cats))
```

```{r percetage-output-read-coverage}
ggplot(reads_coverage) +
  geom_bar(
    aes(x = category, y = percentage_of_input, group = modality, fill = modality),
    stat = "identity") +
  scale_fill_brewer(palette = "Set2") +
  coord_flip()
```


```{r}
sessionInfo()
```

