---
title: "Multi-omics analysis"
output: html_document
---

```{r knitr, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

```{r setup, message=FALSE}
library("phyloseq")
library("RColorBrewer")
library("tidyverse")
library("readxl")
library("data.table")
library("PMA")
datadir <- "../../data/"
curdir <- getwd()
#setwd(curdir)
```

## Obtain data

```{r}
# Sample Information
meas_info <- read_xlsx(
  file.path(datadir, "sample_info/Mapping_File_19Oct2018.xlsx"),
  "Meas", range = cell_cols(1:7))
names(meas_info) <- meas_info[1, ]
meas_info <- meas_info[-1, ] %>%
  rename(Samp_ID = SampID)

interv_levs <- c("NoInterv", "PreDiet", "MidDiet", "PostDiet", 
                 "PreCC", "MidCC","PostCC", "PreAbx", "MidAbx", "PostAbx")
samp <- read_xlsx(
  file.path(datadir, "sample_info/Mapping_File_19Oct2018.xlsx"),
  "Samp", skip = 1) %>%
  mutate(
    Diet_Interval = ifelse(Diet_Interval == "NA", "NoInterv", Diet_Interval),
    CC_Interval = ifelse(CC_Interval == "NA", "NoInterv", CC_Interval),
    Abx_Interval = ifelse(Abx_Interval == "NA", "NoInterv", Abx_Interval),
    Diet_Interval = factor(Diet_Interval, interv_levs),
    CC_Interval = factor(CC_Interval, interv_levs),
    Abx_Interval = factor(Abx_Interval, interv_levs)
  ) %>%
  filter(Samp_Type != "ExtrCont")

meas_info <- meas_info %>% select(Meas_ID, Samp_ID) %>%
  left_join(samp)
dim(meas_info)
rm(samp)
```


```{r}
# Metabolome, load first since the smallest number of subjects
metabolon <- readRDS(file.path(datadir, "metabolome/assayData NA_filtered_and_imputed.rds"))
```


```{r}
# metatranscriptomics aligned to SEED DB
# This is the coverage abundance measure computed as the total number of bp aligned / length of the gene
metat <- fread(file.path(datadir, "metatranscriptomic/counts/abund_seed_len_ratio.csv"))
dim(metat)
```
```{r}
gene_info <- fread(file.path(datadir, "metatranscriptomic/counts/gene_len_seed_raw.csv"))
metat <- metat[, !c("SEED1", "SEED2", "SEED3", "SEED4"), with = FALSE]
metat <- metat[order(GeneID), ]
gene_info <- gene_info[order(GeneID), ]
identical(metat$GeneID, gene_info$GeneID)
dim(metat)
dim(gene_info)
```




```{r, eval =FALSE}
gene_info2 <- metat[, .(GeneID, SEED1, SEED2, SEED3, SEED4)]
identical(gene_info2[order(GeneID)], 
          gene_info[order(GeneID), !c("Length"), with = FALSE])
rm(gene_info2)

metat_align <- fread(file.path(datadir, "metatranscriptomic/counts/abund_seed_bp_aligned.csv"))
metat_align <- metat_align[, !c("SEED1", "SEED2", "SEED3", "SEED4"), with = FALSE]
metat_align <- metat_align[order(GeneID), ]

setcolorder(metat_align, colnames(metat))
identical(colnames(metat), colnames(metat_align))
identical(metat$GeneID, metat_align$GeneID)

```



```{r, eval = FALSE}
# Sanity check for correctness of measures:
id <- sum_err <- mean_err <- len <-  rep(0, 100)
rnd_idx <- sample(1:nrow(metat), 100)
for(k in seq_along(rnd_idx)) {
  i <- rnd_idx[k]
  sum_err[k] <- sum(abs(metat_align[i, -1]/gene_info$Length[i] - metat[i, -1]), na.rm = TRUE)
  len[k] <- sum(!is.na(metat_align[i, -1]))
  mean_err[k] <- sum_err[k]/len[k]
  id[k] <- identical(!is.na(metat_align[i, -1]), !is.na(metat[i, -1]))
}
summary(sum_err)
rm(metat_align)
```

```{r}
new_colnames <- gsub("\\_.*", "", colnames(metat))
head(new_colnames)
colnames(metat) <- new_colnames
```


Filter, sample to the same as in metabolon 

```{r}
meas_info_sub <- meas_info %>%
  filter(Samp_ID %in% colnames(metabolon))
dim(meas_info_sub)
head(meas_info_sub)
```





```{r}
sum(colnames(metat) %in% meas_info_sub$Meas_ID)
```


```{r}
col_names <- colnames(metat)[(colnames(metat) %in% meas_info_sub$Meas_ID)]
col_names <- c("GeneID", col_names)
metat_sub <- metat[, col_names,  with=FALSE]
dim(metat_sub)
```

```{r}
rsum <- rowSums(col_names[, -1], na.rm = TRUE)
summary(rsum)
sum(rsum > 5)
```

```{r}
metat_sub <- metat_sub[rsum > 5, ]
dim(metat_sub)
```


```{r}
sb_id <- meas_info_sub %>%
  filter(Meas_ID %in% colnames(metat_sub)) %>%
  data.frame() %>%
  column_to_rownames("Meas_ID")
sb_id <- sb_id[colnames(metat_sub)[-1], "Subject"]
```


```{r}
presence_thresh <- 1
subject_prevalence <- apply(metat_sub[, -1], 1, function(x) {
  x[is.na(x)] <- 0
  sapply(unique(sb_id), function(y) sum(x[sb_id == y] > presence_thresh))
})
subject_prevalence <- t(subject_prevalence)

num_samples_present_for_subj <- 2
z <- apply(subject_prevalence, 1, function(x) sum(x > num_samples_present_for_subj))
sum(z == 3)
```

```{r}
metat_sub <- metat_sub[z==3, ]
dim(metat_sub)
save(list = c("metat_sub", "metabolon", "meas_info_sub"), file = "multiomics_res.rda")
```

```{r}
gene_info_sub <- gene_info %>%
  filter(GeneID %in% metat_sub$GeneID)
```


```{r}
ps <- readRDS(file.path(datadir, "16S/phyloseq/perturb_physeq_fltr2430_17Jul18.rds"))
ps
```

```{r}
ps <- subset_samples(ps, Subject %in% unique(meas_info_sub$Subject))
ps
```

```{r}
presence_thresh <- 1
subject_prevalence <- apply(as(otu_table(ps), "matrix"), 1, function(x) {
  x[is.na(x)] <- 0
  sapply(unique(sb_id), function(y) sum(x[sb_id == y] > presence_thresh))
})
subject_prevalence <- t(subject_prevalence)

num_samples_present_for_subj <- 2
z <- apply(subject_prevalence, 1, function(x) sum(x > num_samples_present_for_subj))
sum(z == 3)
```

```{r}
ps <- subset_taxa(ps, z == 3)
ps
```


```{r}
save(list = c("metat_sub", "metabolon", "meas_info_sub", "ps"), file = "multiomics_res.rda")
```


## Sparse CCA


```{r}
load("multiomics_res.rda")
```


```{r}
meas_info_sub <- meas_info_sub %>%
  data.frame() %>%
  mutate("MeasID" = Meas_ID) %>%
  column_to_rownames("MeasID")
```


```{r}
Xmetat <- as.matrix(metat_sub[, -1])
Xmetat[is.na(Xmetat)] <- 0
rownames(Xmetat) <- metat_sub$GeneID
Xmetat <- t(Xmetat)
rownames(Xmetat) <- meas_info_sub[rownames(Xmetat), "Samp_ID"]
Xmetat <- asinh(Xmetat)

Xphyseq <- as(otu_table(ps), "matrix")
Xphyseq <- t(Xphyseq)
rownames(Xphyseq) <- meas_info_sub[rownames(Xphyseq), "Samp_ID"]
Xphyseq <- asinh(Xphyseq)

Xmetabol <- t(metabolon)
Xmetabol <- asinh(Xmetabol)
```


```{r}
dim(Xmetabol)
dim(Xphyseq)
dim(Xmetat)
```


```{r}
sampleNames <- rownames(Xmetabol)
head(sampleNames)
```


```{r}
Xmetat <- Xmetat[sampleNames, ]
Xphyseq <- Xphyseq[sampleNames, ]
```

```{r}
sum(rownames(Xphyseq) != rownames(Xmetabol))
sum(rownames(Xmetat) != rownames(Xmetabol))
```

```{r}
dim(Xmetabol)
dim(Xphyseq)
dim(Xmetat)
```


```{r}
spCCA <- MultiCCA.permute(list(Xmetabol, Xphyseq, Xmetat)) 
```



```{r}
save(list = c("metat_sub", "metabolon", "meas_info_sub", "ps", "spCCA", "gene_info_sub"), 
     file = "multiomics_res.rda")
```




```{r}
print(spCCA)
```

```{r}
plot(spCCA)
```



```{r}
spCCA.out <- MultiCCA(list(Xmetabol, Xphyseq, Xmetat), 
                      penalty=spCCA$bestpenalties, ncomponents=1, 
                      ws=spCCA$ws.init)
print(spCCA.out)
```

```{r}
save(list = c("metat_sub", "metabolon", "meas_info_sub", "ps", "spCCA", "spCCA.out"), 
     file = "multiomics_res.rda")
```

```{r}
names(spCCA.out$ws)
```

```{r}
par(mfrow=c(3,1))
PlotCGH(spCCA.out$ws[[1]])
PlotCGH(spCCA.out$ws[[2]])
PlotCGH(spCCA.out$ws[[3]])
```


```{r}
Wmetabol <- Xmetabol[, abs(spCCA.out$ws[[1]][, 1]) > 0]
Wphyseq <- Xphyseq[, abs(spCCA.out$ws[[2]][, 1]) > 0]
Wmetat <- Xmetat[, abs(spCCA.out$ws[[3]][, 1]) > 0]
```


```{r}
ds_id <- c(rep("metabol", sum(abs(spCCA.out$ws[[1]][, 1]) > 0)),
           rep("16S", sum(abs(spCCA.out$ws[[2]][, 1]) > 0)),
           rep("metatrans", sum(abs(spCCA.out$ws[[3]][, 1]) > 0)))

X <- cbind(Wmetabol, Wphyseq, Wmetat)
pca.res <- prcomp(X)
```


```{r}
Y <- pca.res$x[, 1:3] %>%
  data.frame()
Feat <- pca.res$rotation[, 1:3] %>%
  data.frame() 

Feat <- Feat %>%
  mutate(Feat = rownames(Feat), DS = ds_id)


Y <- Y %>%
  mutate(Samp_ID = rownames(Y)) 

Y <- Y %>%
  left_join(meas_info_sub)


  
var.exp <- round(100*pca.res$sdev^2/sum(pca.res$sdev^2), digits = 2)
```

```{r}
ggplot(Y, aes(PC1, PC2)) +
  geom_point(aes(fill = Subject), size = 3, pch = 21, color = "white") +
  geom_point(data = Feat, color = "grey60", size = 2,
             aes(shape = DS, x = PC1*200, y = PC2*200)) +
  coord_fixed() +
  xlab(paste0("PC1 [", var.exp[1], "%]")) +
  ylab(paste0("PC2 [", var.exp[2], "%]"))
```


```{r}
gene_info_sub %>% filter(GeneID %in% colnames(Wmetat))
```


```{r}
ps@tax_table %>%
  data.frame() %>%
  rownames_to_column("Seq_ID") %>%
  filter(Seq_ID %in% colnames(Wphyseq))
```




```{r}
df <- data.frame(Xmetabol[, as.character(colnames(Wmetabol))], check.names = FALSE) %>%
  rownames_to_column("Samp_ID") %>%
  left_join(meas_info_sub)
```


```{r, fig.width=20, fig.height=20}
df %>% 
  gather(key ="metabolite", value = "intensity", `34404`:`62945`) %>%
  filter(Subject != "DBU") %>%
  ggplot(aes(x = Abx_RelDay, y = intensity, color = Subject)) +
  geom_point() + geom_line(aes(group = Subject)) +
  facet_wrap(~metabolite, ncol = 10)
```


```{r}
df <- data.frame(Xmetat[, as.character(colnames(Wmetat))], check.names = FALSE) %>%
  rownames_to_column("Samp_ID") %>%
  left_join(meas_info_sub)
```



```{r, fig.width=20, fig.height=20}
df %>% 
  gather(key ="metabolite", value = "intensity", `34404`:`62945`) %>%
  ggplot(aes(x = Abx_RelDay, y = intensity, color = Subject)) +
  geom_point() + geom_line(aes(group = metabolite)) +
  facet_wrap(~metabolite, ncol = 10)
```




# Seesion Info

```{r}
sessionInfo()
```

